/*
 * Copyright 2023 EMBL - European Bioinformatics Institute
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package uk.ac.ebi.ena.webin.xml.conversion;

import static org.junit.Assert.assertEquals;

import org.json.JSONException;
import org.junit.ComparisonFailure;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;
import org.skyscreamer.jsonassert.Customization;
import org.skyscreamer.jsonassert.JSONAssert;
import org.skyscreamer.jsonassert.JSONCompareMode;
import org.skyscreamer.jsonassert.comparator.CustomComparator;
import org.xmlunit.builder.DiffBuilder;
import org.xmlunit.diff.ComparisonType;
import org.xmlunit.diff.DefaultNodeMatcher;
import org.xmlunit.diff.Diff;
import org.xmlunit.diff.DifferenceEvaluators;
import org.xmlunit.diff.ElementSelectors;

public class EnaSraConverterTest {

  private EnaSraConverter enaSraConverter = new EnaSraConverter();

  @Rule public ExpectedException expectedEx = ExpectedException.none();

  @Test
  public void testConvertSubmissionJsonToXml() {
    final String webinSubmissionJson =
        "{\n"
            + "   \"submission\":{\n"
            + "      \"alias\":\"test-dipayan-205\",\n"
            + "      \"accession\":\"\",\n"
            + "      \"actions\":[\n"
            + "         {\n"
            + "            \"type\":\"ADD\"\n"
            + "         },\n"
            + "         {\n"
            + "            \"type\":\"HOLD\",\n"
            + "            \"holdUntilDate\":\"2022-01-01\"\n"
            + "         }\n"
            + "      ],\n"
            + "      \"attributes\":[\n"
            + "         {\n"
            + "            \"tag\":\"test\",\n"
            + "            \"value\":\"test_val\"\n"
            + "         },\n"
            + "         {\n"
            + "            \"tag\":\"test_1\",\n"
            + "            \"value\":\"test_val_1\"\n"
            + "         }\n"
            + "      ]\n"
            + "   },\n"
            + "   \"samples\":[\n"
            + "      {\n"
            + "         \"alias\":\"sample_alias_110\",\n"
            + "         \"title\":\"testTitle\",\n"
            + "         \"description\":\"testDescription\",\n"
            + "         \"organism\":{\n"
            + "            \"taxonId\":\"9606\"\n"
            + "         },\n"
            + "         \"attributes\":[\n"
            + "            {\n"
            + "               \"tag\":\"XXX\",\n"
            + "               \"value\":\"YYY\"\n"
            + "            },\n"
            + "            {\n"
            + "               \"tag\":\"XXX\",\n"
            + "               \"value\":\"YYY\",\n"
            + "               \"unit\":\"mg\"\n"
            + "            }\n"
            + "         ]\n"
            + "      },\n"
            + "      {\n"
            + "         \"alias\":\"sample_alias_111\",\n"
            + "         \"organism\":{\n"
            + "            \"taxonId\":\"9606\"\n"
            + "         },\n"
            + "         \"attributes\":[\n"
            + "            {\n"
            + "               \"tag\":\"XXX\",\n"
            + "               \"value\":\"YYY\"\n"
            + "            },\n"
            + "            {\n"
            + "               \"tag\":\"XXX\",\n"
            + "               \"value\":\"YYY\",\n"
            + "               \"unit\":\"mg\"\n"
            + "            }\n"
            + "         ]\n"
            + "      }\n"
            + "   ]\n"
            + "}";

    assertXml(
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
            + "\n"
            + "<WEBIN>\n"
            + "  <SUBMISSION alias=\"test-dipayan-205\" accession=\"\">\n"
            + "    <ACTIONS>\n"
            + "      <ACTION>\n"
            + "        <ADD/>\n"
            + "      </ACTION>\n"
            + "      <ACTION>\n"
            + "        <HOLD HoldUntilDate=\"2022-01-01\"/>\n"
            + "      </ACTION>\n"
            + "    </ACTIONS>\n"
            + "    <SUBMISSION_ATTRIBUTES>\n"
            + "      <SUBMISSION_ATTRIBUTE>\n"
            + "        <TAG>test</TAG>\n"
            + "        <VALUE>test_val</VALUE>\n"
            + "      </SUBMISSION_ATTRIBUTE>\n"
            + "      <SUBMISSION_ATTRIBUTE>\n"
            + "        <TAG>test_1</TAG>\n"
            + "        <VALUE>test_val_1</VALUE>\n"
            + "      </SUBMISSION_ATTRIBUTE>\n"
            + "    </SUBMISSION_ATTRIBUTES>\n"
            + "  </SUBMISSION>\n"
            + "  <SAMPLE_SET>\n"
            + "    <SAMPLE alias=\"sample_alias_110\">\n"
            + "      <TITLE>testTitle</TITLE>\n"
            + "      <DESCRIPTION>testDescription</DESCRIPTION>\n"
            + "      <SAMPLE_NAME>\n"
            + "        <TAXON_ID>9606</TAXON_ID>\n"
            + "      </SAMPLE_NAME>\n"
            + "      <SAMPLE_ATTRIBUTES>\n"
            + "        <SAMPLE_ATTRIBUTE>\n"
            + "          <TAG>XXX</TAG>\n"
            + "          <VALUE>YYY</VALUE>\n"
            + "        </SAMPLE_ATTRIBUTE>\n"
            + "        <SAMPLE_ATTRIBUTE>\n"
            + "          <TAG>XXX</TAG>\n"
            + "          <VALUE>YYY</VALUE>\n"
            + "          <UNITS>mg</UNITS>\n"
            + "        </SAMPLE_ATTRIBUTE>\n"
            + "      </SAMPLE_ATTRIBUTES>\n"
            + "    </SAMPLE>\n"
            + "    <SAMPLE alias=\"sample_alias_111\">\n"
            + "      <SAMPLE_NAME>\n"
            + "        <TAXON_ID>9606</TAXON_ID>\n"
            + "      </SAMPLE_NAME>\n"
            + "      <SAMPLE_ATTRIBUTES>\n"
            + "        <SAMPLE_ATTRIBUTE>\n"
            + "          <TAG>XXX</TAG>\n"
            + "          <VALUE>YYY</VALUE>\n"
            + "        </SAMPLE_ATTRIBUTE>\n"
            + "        <SAMPLE_ATTRIBUTE>\n"
            + "          <TAG>XXX</TAG>\n"
            + "          <VALUE>YYY</VALUE>\n"
            + "          <UNITS>mg</UNITS>\n"
            + "        </SAMPLE_ATTRIBUTE>\n"
            + "      </SAMPLE_ATTRIBUTES>\n"
            + "    </SAMPLE>\n"
            + "  </SAMPLE_SET>\n"
            + "</WEBIN>\n",
        enaSraConverter.convertSubmissionJsonToXml(webinSubmissionJson));
  }

  @Test
  public void testConvertSubmissionXmlToJson() throws JSONException {
    String submissionXml =
        "<SUBMISSION_SET>\n"
            + "  <SUBMISSION accession=\"ERA3206932\" alias=\"ena-SUBMISSION-SSI-11-01-2021-14:27:35:627-1\" center_name=\"Statens Serum Institut\" submission_date=\"2021-01-11T14:28:34.059Z\">\n"
            + "    <IDENTIFIERS>\n"
            + "      <PRIMARY_ID>ERA3206932</PRIMARY_ID>\n"
            + "      <SUBMITTER_ID namespace=\"Statens Serum Institut\">ena-SUBMISSION-SSI-11-01-2021-14:27:35:627-1</SUBMITTER_ID>\n"
            + "    </IDENTIFIERS>\n"
            + "    <TITLE>Submitted by Statens Serum Institut on 11-JAN-2021</TITLE>\n"
            + "    <SUBMISSION_LINKS>\n"
            + "      <SUBMISSION_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>ENA-FASTQ-FILES</DB>\n"
            + "          <ID><![CDATA[https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERA3206932&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes]]></ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SUBMISSION_LINK>\n"
            + "      <SUBMISSION_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>ENA-SUBMITTED-FILES</DB>\n"
            + "          <ID><![CDATA[https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERA3206932&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format]]></ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SUBMISSION_LINK>\n"
            + "    </SUBMISSION_LINKS>\n"
            + "    <SUBMISSION_ATTRIBUTES>\n"
            + "      <SUBMISSION_ATTRIBUTE>\n"
            + "        <TAG>test</TAG>\n"
            + "        <VALUE>test_val</VALUE>\n"
            + "      </SUBMISSION_ATTRIBUTE>\n"
            + "      <SUBMISSION_ATTRIBUTE>\n"
            + "        <TAG>test_1</TAG>\n"
            + "        <VALUE>test_val_1</VALUE>\n"
            + "      </SUBMISSION_ATTRIBUTE>\n"
            + "    </SUBMISSION_ATTRIBUTES>\n"
            + "  </SUBMISSION>\n"
            + "</SUBMISSION_SET>";

    assertJson(
        "{\n"
            + "  \"alias\" : \"ena-SUBMISSION-SSI-11-01-2021-14:27:35:627-1\",\n"
            + "  \"accession\" : \"ERA3206932\",\n"
            + "  \"centerName\" : \"Statens Serum Institut\",\n"
            + "  \"submissionDate\" : \"2021-01-11T14:28:34.059Z\",\n"
            + "  \"title\" : \"Submitted by Statens Serum Institut on 11-JAN-2021\",\n"
            + "  \"identifiers\" : { },\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"test\",\n"
            + "    \"value\" : \"test_val\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"test_1\",\n"
            + "    \"value\" : \"test_val_1\"\n"
            + "  } ],\n"
            + "  \"links\" : [ {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-FASTQ-FILES\",\n"
            + "      \"id\" : \"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERA3206932&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SUBMITTED-FILES\",\n"
            + "      \"id\" : \"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERA3206932&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format\"\n"
            + "    }\n"
            + "  } ]\n"
            + "}",
        enaSraConverter.convertSubmissionXmlToJson(submissionXml));
  }

  @Test
  public void testConvertSubmissionJsonWithHoldActionToXml() {
    String submissionJson =
        "{\n"
            + "   \"submission\":{\n"
            + "      \"alias\":\"subs_hold_test_1\",\n"
            + "      \"accession\":\"\",\n"
            + "      \"actions\":[\n"
            + "         {\n"
            + "            \"type\":\"MODIFY\"\n"
            + "         },\n"
            + "         {\n"
            + "            \"type\":\"HOLD\",\n"
            + "            \"target\":\"ERS17064574\",\n"
            + "            \"holdUntilDate\":\"2027-01-01\"\n"
            + "         }\n"
            + "      ]\n"
            + "   }\n"
            + "}";

    assertXml(
        "<WEBIN>\n"
            + "  <SUBMISSION alias=\"subs_hold_test_1\" accession=\"\">\n"
            + "    <ACTIONS>\n"
            + "      <ACTION>\n"
            + "        <MODIFY/>\n"
            + "      </ACTION>\n"
            + "      <ACTION>\n"
            + "        <HOLD HoldUntilDate=\"2027-01-01\" target=\"ERS17064574\"/>\n"
            + "      </ACTION>\n"
            + "    </ACTIONS>\n"
            + "  </SUBMISSION>\n"
            + "</WEBIN>",
        enaSraConverter.convertSubmissionJsonToXml(submissionJson));
  }

  @Test
  public void testConvertSubmissionJsonWithReleaseActionToXml() {
    String submissionJson =
        "{\n"
            + "   \"submission\":{\n"
            + "      \"alias\":\"subs_hold_test_1\",\n"
            + "      \"accession\":\"\",\n"
            + "      \"actions\":[\n"
            + "         {\n"
            + "            \"type\":\"RELEASE\",\n"
            + "            \"target\":\"ERS17064574\"\n"
            + "         }\n"
            + "      ]\n"
            + "   }\n"
            + "}";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<WEBIN>\n"
            + "  <SUBMISSION alias=\"subs_hold_test_1\" accession=\"\">\n"
            + "    <ACTIONS>\n"
            + "      <ACTION>\n"
            + "        <RELEASE target=\"ERS17064574\"/>\n"
            + "      </ACTION>\n"
            + "    </ACTIONS>\n"
            + "  </SUBMISSION>\n"
            + "</WEBIN>",
        enaSraConverter.convertSubmissionJsonToXml(submissionJson));
  }

  @Test
  public void testConvertSubmission_MultipleObjects_JsonToXml() {
    final String webinSubmissionJson =
        "{\n"
            + "   \"submission\":{\n"
            + "      \"alias\":\"subs_test-alias-117\",\n"
            + "      \"accession\":\"\",\n"
            + "      \"actions\":[\n"
            + "         {\n"
            + "            \"type\":\"ADD\"\n"
            + "         },\n"
            + "         {\n"
            + "            \"type\":\"HOLD\",\n"
            + "            \"holdUntilDate\":\"2023-01-01\"\n"
            + "         }\n"
            + "      ],\n"
            + "      \"attributes\":[\n"
            + "         {\n"
            + "            \"tag\":\"test_tag\",\n"
            + "            \"value\":\"test_val\"\n"
            + "         },\n"
            + "         {\n"
            + "            \"tag\":\"test_tag_1\",\n"
            + "            \"value\":\"test_val_1\"\n"
            + "         }\n"
            + "      ]\n"
            + "   },\n"
            + "   \"analyses\":[\n"
            + "      {\n"
            + "         \"alias\":\"analysis_3\",\n"
            + "         \"title\":\"test\",\n"
            + "         \"description\":\"test\",\n"
            + "         \"study\":{\n"
            + "            \"accession\":\"ERP000004\"\n"
            + "         },\n"
            + "         \"samples\":[\n"
            + "            {\n"
            + "               \"alias\":\"test_sample_1\"\n"
            + "            }\n"
            + "         ],\n"
            + "         \"analysisType\":\"ASSEMBLY_GRAPH\",\n"
            + "         \"assemblies\":[\n"
            + "            {\n"
            + "               \"assembly\":{\n"
            + "                  \"accession\":\"GCA_012345.1\"\n"
            + "               }\n"
            + "            },\n"
            + "            {\n"
            + "               \"assembly\":{\n"
            + "                  \"accession\":\"GCA_012346.1\"\n"
            + "               }\n"
            + "            },\n"
            + "            {\n"
            + "               \"assembly\":{\n"
            + "                  \"accession\":\"GCA_012347.1\"\n"
            + "               }\n"
            + "            }\n"
            + "         ],\n"
            + "         \"files\":[\n"
            + "            {\n"
            + "               \"fileName\":\"/webin-cli/genome/SPSP3996252-2/SPSP3996252-3.gfa.gz\",\n"
            + "               \"fileType\":\"gfa\",\n"
            + "               \"checksumMethod\":\"MD5\",\n"
            + "               \"checksum\":\"2ba69754fee276fd62d7310068d3e07d\"\n"
            + "            }\n"
            + "         ],\n"
            + "         \"attributes\":[\n"
            + "            {\n"
            + "               \"tag\":\"test_attr\",\n"
            + "               \"value\":\"test_val\"\n"
            + "            }\n"
            + "         ]\n"
            + "      }\n"
            + "   ],\n"
            + "   \"runs\":[\n"
            + "      {\n"
            + "         \"alias\":\"run_alias_4\",\n"
            + "         \"identifiers\":null,\n"
            + "         \"centerName\":\"EBI\",\n"
            + "         \"brokerName\":\"EBI\",\n"
            + "         \"title\":\"test\",\n"
            + "         \"description\":\"test\",\n"
            + "         \"experiment\":{\n"
            + "            \"alias\":\"ena-EXPERIMENT-2\"\n"
            + "         },\n"
            + "         \"instrumentPlatform\":\"ILLUMINA\",\n"
            + "         \"instrumentModel\":\"Illumina MiSeq\",\n"
            + "         \"files\":[\n"
            + "            {\n"
            + "               \"fileName\":\"/webin-cli/genome/SPSP3996252-2/SPSP3996252-2.bam\",\n"
            + "               \"fileType\":\"bam\",\n"
            + "               \"checksumMethod\":\"MD5\",\n"
            + "               \"checksum\":\"2ba69754fee276fd62d7310068d3e07d\",\n"
            + "               \"qualityScoringSystem\":\"phred\"\n"
            + "            }\n"
            + "         ],\n"
            + "         \"attributes\":[\n"
            + "            {\n"
            + "               \"tag\":\"test_tag\",\n"
            + "               \"value\":\"test_val\"\n"
            + "            },\n"
            + "            {\n"
            + "               \"tag\":\"test_tag_1\",\n"
            + "               \"value\":\"test_val_1\"\n"
            + "            }\n"
            + "         ]\n"
            + "      }\n"
            + "   ],\n"
            + "   \"experiments\":[\n"
            + "      {\n"
            + "         \"alias\":\"ena-EXPERIMENT-2\",\n"
            + "         \"identifiers\":null,\n"
            + "         \"centerName\":\"EBI\",\n"
            + "         \"title\":\"Illumina MiSeq sequencing\",\n"
            + "         \"study\":{\n"
            + "            \"accession\":\"ERP121346\"\n"
            + "         },\n"
            + "         \"samples\":[\n"
            + "            {\n"
            + "               \"alias\":\" test_sample_1\"\n"
            + "            }\n"
            + "         ],\n"
            + "         \"designDescription\":\"\",\n"
            + "         \"libraryDescriptor\":{\n"
            + "            \"libraryName\":\"unspecified\",\n"
            + "            \"libraryStrategy\":\"AMPLICON\",\n"
            + "            \"librarySource\":\"GENOMIC\",\n"
            + "            \"librarySelection\":\"PCR\",\n"
            + "            \"libraryLayout\":\"PAIRED\"\n"
            + "         },\n"
            + "         \"instrumentPlatform\":\"ILLUMINA\",\n"
            + "         \"instrumentModel\":\"Illumina MiSeq\",\n"
            + "         \"attributes\":[\n"
            + "            {\n"
            + "               \"tag\":\"test_tag\",\n"
            + "               \"value\":\"test_val\"\n"
            + "            },\n"
            + "            {\n"
            + "               \"tag\":\"test_tag_1\",\n"
            + "               \"value\":\"test_val_1\"\n"
            + "            }\n"
            + "         ]\n"
            + "      }\n"
            + "   ],\n"
            + "   \"samples\":[\n"
            + "      {\n"
            + "         \"alias\":\"test_sample_1\",\n"
            + "         \"title\":\"testTitle\",\n"
            + "         \"organism\":{\n"
            + "            \"taxonId\":\"9606\"\n"
            + "         },\n"
            + "         \"attributes\":[\n"
            + "            {\n"
            + "               \"tag\":\"ena-checklist\",\n"
            + "               \"value\":\"ERC000011\"\n"
            + "            },\n"
            + "            {\n"
            + "               \"tag\":\"test_tag_1\",\n"
            + "               \"value\":\"test_val_2\",\n"
            + "               \"unit\":\"test_unit_1\"\n"
            + "            }\n"
            + "         ]\n"
            + "      }\n"
            + "   ]\n"
            + "}";
  }

  @Test
  public void testConvertInvalidSubmissionJsonToXml() {
    expectedEx.expect(WebinXmlConversionException.class);

    final String webinSubmissionJson =
        "{\n"
            + "   \"submission\":{\n"
            + "      \"alias\":\"test-dipayan-137\",\n"
            + "      \"accession\":\"\",\n"
            + "      \"actions\":[\n"
            + "         {\n"
            + "            \"type\":\"ADD\"\n"
            + "         }\n"
            + "      ],\n"
            + "      \"attributes\":[\n"
            + "         {\n"
            + "            \"tag\":\"test\",\n"
            + "            \"value\":\"test_val\"\n"
            + "         },\n"
            + "         {\n"
            + "            \"tag\":\"test_1\",\n"
            + "            \"value\":\"test_val_1\"\n"
            + "         }\n"
            + "      ]\n"
            + "   },\n"
            + "   \"samples\":[\n"
            + "      {\n"
            + "         \"alias\":\"sample_alias_29\",\n"
            + "         \"INVALID\":\"ERS000001\",\n"
            + "         \"taxId\":\"9606\",\n"
            + "         \"attributes\":[\n"
            + "            {\n"
            + "               \"tag\":\"XXX\",\n"
            + "               \"value\":\"YYY\"\n"
            + "            },\n"
            + "            {\n"
            + "               \"tag\":\"XXX\",\n"
            + "               \"value\":\"YYY\",\n"
            + "               \"unit\":\"mg\"\n"
            + "            }\n"
            + "         ]\n"
            + "      }\n"
            + "   ]\n"
            + "}";

    enaSraConverter.convertSubmissionJsonToXml(webinSubmissionJson);
  }

  @Test
  public void testConvertReceiptJsonToXml() {
    final String receiptJson =
        "{\n"
            + "  \"success\": true,\n"
            + "  \"receiptDate\": \"2022-08-19T17:25:52.635+01:00\",\n"
            + "  \"submissionFile\": \"submission-EMBL-EBI_1660926352635.xml\",\n"
            + "  \"samples\": [\n"
            + "    {\n"
            + "      \"alias\": \"sample_alias_13\",\n"
            + "      \"accession\": \"ERS11837633\",\n"
            + "      \"externalAccession\": {\n"
            + "        \"id\": \"SAMEA110675512\",\n"
            + "        \"db\": \"biosample\"\n"
            + "      }\n"
            + "    },\n"
            + "    {\n"
            + "      \"alias\": \"sample_alias_14\",\n"
            + "      \"accession\": \"ERS11837634\",\n"
            + "      \"externalAccession\": {\n"
            + "        \"id\": \"SAMEA110675513\",\n"
            + "        \"db\": \"biosample\"\n"
            + "      }\n"
            + "    }\n"
            + "  ],\n"
            + "  \"submission\": {\n"
            + "    \"alias\": \"test-dipayan-129\",\n"
            + "    \"accession\": \"ERA12880855\"\n"
            + "  },\n"
            + "  \"messages\": {\n"
            + "    \"error\": [\n"
            + "      \"Any.\"\n"
            + "    ],\n"
            + "    \"info\": [\n"
            + "      \"All objects in this submission are set to private status (HOLD).\"\n"
            + "    ]\n"
            + "  },\n"
            + "  \"actions\": [\n"
            + "    \"ADD\",\n"
            + "    \"HOLD\"\n"
            + "  ]\n"
            + "}";

    assertXml(
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
            + "\n"
            + "<RECEIPT success=\"true\" receiptDate=\"2022-08-19T17:25:52.635+01:00\">\n"
            + "  <SAMPLE alias=\"sample_alias_13\" accession=\"ERS11837633\">\n"
            + "    <EXT_ID accession=\"SAMEA110675512\" type=\"biosample\"/>\n"
            + "  </SAMPLE>\n"
            + "  <SAMPLE alias=\"sample_alias_14\" accession=\"ERS11837634\">\n"
            + "    <EXT_ID accession=\"SAMEA110675513\" type=\"biosample\"/>\n"
            + "  </SAMPLE>\n"
            + "  <SUBMISSION alias=\"test-dipayan-129\" accession=\"ERA12880855\"/>\n"
            + "  <MESSAGES>\n"
            + "    <INFO>All objects in this submission are set to private status (HOLD).</INFO>\n"
            + "    <ERROR>Any.</ERROR>\n"
            + "  </MESSAGES>\n"
            + "  <ACTIONS>ADD</ACTIONS>\n"
            + "  <ACTIONS>HOLD</ACTIONS>\n"
            + "</RECEIPT>\n",
        enaSraConverter.convertReceiptJsonToXml(receiptJson));
  }

  @Test
  public void testConvertReceiptXmlToJsonWithMessagesOfSameType() throws JSONException {
    final String receiptXml =
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
            + "  <?xml-stylesheet type=\"text/xsl\" href=\"receipt.xsl\"?>\n"
            + "    <RECEIPT receiptDate=\"2022-08-19T17:25:52.635+01:00\" submissionFile=\"submission-EMBL-EBI_1660926352635.xml\" success=\"true\">\n"
            + "      <SAMPLE accession=\"ERS11837633\" alias=\"sample_alias_13\" status=\"PRIVATE\" holdUntilDate=\"2022-01-01Z\">\n"
            + "        <EXT_ID accession=\"SAMEA110675512\" type=\"biosample\"/>\n"
            + "      </SAMPLE>\n"
            + "      <SAMPLE accession=\"ERS11837634\" alias=\"sample_alias_14\" status=\"PRIVATE\" holdUntilDate=\"2022-01-01Z\">\n"
            + "        <EXT_ID accession=\"SAMEA110675513\" type=\"biosample\"/>\n"
            + "      </SAMPLE>\n"
            + "      <SUBMISSION accession=\"ERA12880855\" alias=\"test-dipayan-129\"/>\n"
            + "      <MESSAGES>\n"
            + "        <INFO>All objects in this submission are set to private status (HOLD).</INFO>\n"
            + "        <INFO>Any.</INFO>\n"
            + "        <ERROR>Any.</ERROR>\n"
            + "      </MESSAGES>\n"
            + "      <ACTIONS>ADD</ACTIONS>\n"
            + "      <ACTIONS>HOLD</ACTIONS>\n"
            + "    </RECEIPT>";

    assertJson(
        "{\n"
            + "  \"success\": true,\n"
            + "  \"receiptDate\": \"2022-08-19T17:25:52.635+01:00\",\n"
            + "  \"samples\": [\n"
            + "    {\n"
            + "      \"alias\": \"sample_alias_13\",\n"
            + "      \"accession\": \"ERS11837633\",\n"
            + "      \"status\": \"PRIVATE\",\n"
            + "      \"holdUntilDate\": \"2022-01-01Z\",\n"
            + "      \"externalAccession\": {\n"
            + "        \"id\": \"SAMEA110675512\",\n"
            + "        \"db\": \"biosample\"\n"
            + "      }\n"
            + "    },\n"
            + "    {\n"
            + "      \"alias\": \"sample_alias_14\",\n"
            + "      \"accession\": \"ERS11837634\",\n"
            + "      \"status\": \"PRIVATE\",\n"
            + "      \"holdUntilDate\": \"2022-01-01Z\",\n"
            + "      \"externalAccession\": {\n"
            + "        \"id\": \"SAMEA110675513\",\n"
            + "        \"db\": \"biosample\"\n"
            + "      }\n"
            + "    }\n"
            + "  ],\n"
            + "  \"submission\": {\n"
            + "    \"alias\": \"test-dipayan-129\",\n"
            + "    \"accession\": \"ERA12880855\"\n"
            + "  },\n"
            + "  \"messages\": {\n"
            + "    \"info\": [\n"
            + "      \"All objects in this submission are set to private status (HOLD).\",\n"
            + "      \"Any.\"\n"
            + "    ],\n"
            + "    \"error\": [\n"
            + "      \"Any.\"\n"
            + "    ]\n"
            + "  },\n"
            + "  \"actions\": [\n"
            + "    \"ADD\",\n"
            + "    \"HOLD\"\n"
            + "  ]\n"
            + "}",
        enaSraConverter.convertReceiptXmlToJson(receiptXml));
  }

  @Test
  public void testConvertReceiptXmlToJson() throws JSONException {
    final String receiptXml =
        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
            + "  <?xml-stylesheet type=\"text/xsl\" href=\"receipt.xsl\"?>\n"
            + "    <RECEIPT receiptDate=\"2022-08-19T17:25:52.635+01:00\" submissionFile=\"submission-EMBL-EBI_1660926352635.xml\" success=\"true\">\n"
            + "      <SAMPLE accession=\"ERS11837633\" alias=\"sample_alias_13\" status=\"PRIVATE\" holdUntilDate=\"2022-01-01Z\">\n"
            + "        <EXT_ID accession=\"SAMEA110675512\" type=\"biosample\"/>\n"
            + "      </SAMPLE>\n"
            + "      <SAMPLE accession=\"ERS11837634\" alias=\"sample_alias_14\" status=\"PRIVATE\" holdUntilDate=\"2022-01-01Z\">\n"
            + "        <EXT_ID accession=\"SAMEA110675513\" type=\"biosample\"/>\n"
            + "      </SAMPLE>\n"
            + "      <SUBMISSION accession=\"ERA12880855\" alias=\"test-dipayan-129\"/>\n"
            + "      <MESSAGES>\n"
            + "        <INFO>All objects in this submission are set to private status (HOLD).</INFO>\n"
            + "        <ERROR>Any.</ERROR>\n"
            + "      </MESSAGES>\n"
            + "      <ACTIONS>ADD</ACTIONS>\n"
            + "      <ACTIONS>HOLD</ACTIONS>\n"
            + "    </RECEIPT>";

    assertJson(
        "{\n"
            + "  \"success\": true,\n"
            + "  \"receiptDate\": \"2022-08-19T17:25:52.635+01:00\",\n"
            + "  \"samples\": [\n"
            + "    {\n"
            + "      \"alias\": \"sample_alias_13\",\n"
            + "      \"accession\": \"ERS11837633\",\n"
            + "      \"status\": \"PRIVATE\",\n"
            + "      \"holdUntilDate\": \"2022-01-01Z\",\n"
            + "      \"externalAccession\": {\n"
            + "        \"id\": \"SAMEA110675512\",\n"
            + "        \"db\": \"biosample\"\n"
            + "      }\n"
            + "    },\n"
            + "    {\n"
            + "      \"alias\": \"sample_alias_14\",\n"
            + "      \"accession\": \"ERS11837634\",\n"
            + "      \"status\": \"PRIVATE\",\n"
            + "      \"holdUntilDate\": \"2022-01-01Z\",\n"
            + "      \"externalAccession\": {\n"
            + "        \"id\": \"SAMEA110675513\",\n"
            + "        \"db\": \"biosample\"\n"
            + "      }\n"
            + "    }\n"
            + "  ],\n"
            + "  \"submission\": {\n"
            + "    \"alias\": \"test-dipayan-129\",\n"
            + "    \"accession\": \"ERA12880855\"\n"
            + "  },\n"
            + "  \"messages\": {\n"
            + "    \"info\": [\n"
            + "      \"All objects in this submission are set to private status (HOLD).\"\n"
            + "    ],\n"
            + "    \"error\": [\n"
            + "      \"Any.\"\n"
            + "    ]\n"
            + "  },\n"
            + "  \"actions\": [\n"
            + "    \"ADD\",\n"
            + "    \"HOLD\"\n"
            + "  ]\n"
            + "}",
        enaSraConverter.convertReceiptXmlToJson(receiptXml));
  }

  @Test
  public void testConvertSampleXmlToJson() throws JSONException {
    final String sampleXml =
        "<SAMPLE_SET>\n"
            + "  <SAMPLE accession=\"ERS11358981\" alias=\"test-dipayan-5\" center_name=\"EMBL-EBI\">\n"
            + "    <IDENTIFIERS>\n"
            + "      <PRIMARY_ID>ERS11358981</PRIMARY_ID>\n"
            + "      <SUBMITTER_ID namespace=\"BioSample\">SAMEA11000000</SUBMITTER_ID>\n"
            + "    </IDENTIFIERS>\n"
            + "    <TITLE>Test Sample</TITLE>\n"
            + "    <SAMPLE_NAME>\n"
            + "      <TAXON_ID>9606</TAXON_ID>\n"
            + "      <SCIENTIFIC_NAME>Homo sapiens</SCIENTIFIC_NAME>\n"
            + "      <COMMON_NAME>human</COMMON_NAME>\n"
            + "    </SAMPLE_NAME>\n"
            + "    <SAMPLE_ATTRIBUTES>\n"
            + "      <SAMPLE_ATTRIBUTE>\n"
            + "        <TAG>project name</TAG>\n"
            + "        <VALUE>Tara Oceans expedition (2009-2013)</VALUE>\n"
            + "      </SAMPLE_ATTRIBUTE>\n"
            + "      <SAMPLE_ATTRIBUTE>\n"
            + "        <TAG>ENA-CHECKLIST</TAG>\n"
            + "        <VALUE>ERC000030</VALUE>\n"
            + "      </SAMPLE_ATTRIBUTE>\n"
            + "    </SAMPLE_ATTRIBUTES>\n"
            + "  </SAMPLE>\n"
            + "</SAMPLE_SET>";

    assertJson(
        "{\n"
            + "  \"alias\": \"test-dipayan-5\",\n"
            + "  \"accession\": \"ERS11358981\",\n"
            + "  \"centerName\": \"EMBL-EBI\",\n"
            + "  \"title\": \"Test Sample\",\n"
            + "  \"organism\": {\n"
            + "    \"taxonId\": \"9606\",\n"
            + "    \"commonName\": \"human\",\n"
            + "    \"scientificName\": \"Homo sapiens\"\n"
            + "  },\n"
            + "  \"attributes\": [\n"
            + "    {\n"
            + "      \"tag\": \"project name\",\n"
            + "      \"value\": \"Tara Oceans expedition (2009-2013)\"\n"
            + "    },\n"
            + "    {\n"
            + "      \"tag\": \"ENA-CHECKLIST\",\n"
            + "      \"value\": \"ERC000030\"\n"
            + "    }\n"
            + "  ]\n"
            + "}",
        enaSraConverter.convertSampleXmlToJson(sampleXml));
  }

  @Test
  public void testConvertSampleXmlToJsonWithIdentifiersMissingIds() throws JSONException {
    final String sampleXml =
        "<SAMPLE_SET>\n"
            + "  <SAMPLE center_name=\"Coriell\" alias=\"GM12096\" accession=\"SRS001264\">\n"
            + "    <IDENTIFIERS>\n"
            + "      <SECONDARY_ID>SRS001264</SECONDARY_ID>\n"
            + "      <EXTERNAL_ID namespace=\"BioSample\"></EXTERNAL_ID>\n"
            + "      <EXTERNAL_ID namespace=\"Coriell\"></EXTERNAL_ID>\n"
            + "      <EXTERNAL_ID namespace=\"Coriell\">NA12096</EXTERNAL_ID>\n"
            + "    </IDENTIFIERS>\n"
            + "    <TITLE>GM12096 male unaffected Ceph/utah pedigree 1332 sample</TITLE>\n"
            + "    <SAMPLE_NAME>\n"
            + "      <TAXON_ID>9606</TAXON_ID>\n"
            + "      <SCIENTIFIC_NAME>Homo sapiens</SCIENTIFIC_NAME>\n"
            + "    </SAMPLE_NAME>\n"
            + "    <DESCRIPTION>Paternal Grandfather</DESCRIPTION>\n"
            + "    <SAMPLE_LINKS>\n"
            + "      <SAMPLE_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>pubmed</DB>\n"
            + "          <ID>9039096</ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SAMPLE_LINK>\n"
            + "      <SAMPLE_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>pubmed</DB>\n"
            + "          <ID>18451141</ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SAMPLE_LINK>\n"
            + "      <SAMPLE_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>pubmed</DB>\n"
            + "          <ID>20497546</ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SAMPLE_LINK>\n"
            + "    </SAMPLE_LINKS>\n"
            + "    <SAMPLE_ATTRIBUTES>\n"
            + "      <SAMPLE_ATTRIBUTE>\n"
            + "        <TAG>tag</TAG>\n"
            + "        <VALUE>Ceph/utah pedigree 1332</VALUE>\n"
            + "      </SAMPLE_ATTRIBUTE>\n"
            + "      <SAMPLE_ATTRIBUTE>\n"
            + "        <TAG>tag</TAG>\n"
            + "        <VALUE>International hapmap project - ceph utah residents with ancestry from northern and western europe</VALUE>\n"
            + "      </SAMPLE_ATTRIBUTE>\n"
            + "    </SAMPLE_ATTRIBUTES>\n"
            + "  </SAMPLE>\n"
            + "</SAMPLE_SET>\n";

    assertJson(
        "{\n"
            + "  \"alias\": \"GM12096\",\n"
            + "  \"accession\": \"SRS001264\",\n"
            + "  \"identifiers\": {\n"
            + "    \"externalAccessions\": [\n"
            + "      {\n"
            + "        \"db\": \"BioSample\"\n"
            + "      },\n"
            + "      {\n"
            + "        \"db\": \"Coriell\"\n"
            + "      },\n"
            + "      {\n"
            + "        \"db\": \"Coriell\",\n"
            + "        \"id\": \"NA12096\"\n"
            + "      }\n"
            + "    ],\n"
            + "    \"secondaryAccessions\": [\n"
            + "      \"SRS001264\"\n"
            + "    ]\n"
            + "  },\n"
            + "  \"centerName\": \"Coriell\",\n"
            + "  \"title\": \"GM12096 male unaffected Ceph/utah pedigree 1332 sample\",\n"
            + "  \"description\": \"Paternal Grandfather\",\n"
            + "  \"organism\": {\n"
            + "    \"taxonId\": \"9606\",\n"
            + "    \"scientificName\": \"Homo sapiens\"\n"
            + "  },\n"
            + "  \"attributes\": [\n"
            + "    {\n"
            + "      \"tag\": \"tag\",\n"
            + "      \"value\": \"Ceph/utah pedigree 1332\"\n"
            + "    },\n"
            + "    {\n"
            + "      \"tag\": \"tag\",\n"
            + "      \"value\": \"International hapmap project - ceph utah residents with ancestry from northern and western europe\"\n"
            + "    }\n"
            + "  ],\n"
            + "  \"links\": [\n"
            + "    {\n"
            + "      \"xrefLink\": {\n"
            + "        \"db\": \"pubmed\",\n"
            + "        \"id\": \"9039096\"\n"
            + "      }\n"
            + "    },\n"
            + "    {\n"
            + "      \"xrefLink\": {\n"
            + "        \"db\": \"pubmed\",\n"
            + "        \"id\": \"18451141\"\n"
            + "      }\n"
            + "    },\n"
            + "    {\n"
            + "      \"xrefLink\": {\n"
            + "        \"db\": \"pubmed\",\n"
            + "        \"id\": \"20497546\"\n"
            + "      }\n"
            + "    }\n"
            + "  ]\n"
            + "}",
        enaSraConverter.convertSampleXmlToJson(sampleXml));
  }

  @Test
  public void testConvertSampleXmlToJsonWithXrefLinks() throws JSONException {
    final String sampleXml =
        "<SAMPLE_SET>\n"
            + "  <SAMPLE center_name=\"Coriell\" alias=\"GM12096\" accession=\"SRS001264\">\n"
            + "    <IDENTIFIERS>\n"
            + "      <SECONDARY_ID>SRS001264</SECONDARY_ID>\n"
            + "      <EXTERNAL_ID namespace=\"BioSample\">SAMN00801146</EXTERNAL_ID>\n"
            + "      <EXTERNAL_ID namespace=\"Coriell\">GM12096</EXTERNAL_ID>\n"
            + "      <EXTERNAL_ID namespace=\"Coriell\">NA12096</EXTERNAL_ID>\n"
            + "    </IDENTIFIERS>\n"
            + "    <TITLE>GM12096 male unaffected Ceph/utah pedigree 1332 sample</TITLE>\n"
            + "    <SAMPLE_NAME>\n"
            + "      <TAXON_ID>9606</TAXON_ID>\n"
            + "      <SCIENTIFIC_NAME>Homo sapiens</SCIENTIFIC_NAME>\n"
            + "    </SAMPLE_NAME>\n"
            + "    <DESCRIPTION>Paternal Grandfather</DESCRIPTION>\n"
            + "    <SAMPLE_LINKS>\n"
            + "      <SAMPLE_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>pubmed</DB>\n"
            + "          <ID>9039096</ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SAMPLE_LINK>\n"
            + "      <SAMPLE_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>pubmed</DB>\n"
            + "          <ID>18451141</ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SAMPLE_LINK>\n"
            + "      <SAMPLE_LINK>\n"
            + "        <XREF_LINK>\n"
            + "          <DB>pubmed</DB>\n"
            + "          <ID>20497546</ID>\n"
            + "        </XREF_LINK>\n"
            + "      </SAMPLE_LINK>\n"
            + "    </SAMPLE_LINKS>\n"
            + "    <SAMPLE_ATTRIBUTES>\n"
            + "      <SAMPLE_ATTRIBUTE>\n"
            + "        <TAG>tag</TAG>\n"
            + "        <VALUE>Ceph/utah pedigree 1332</VALUE>\n"
            + "      </SAMPLE_ATTRIBUTE>\n"
            + "      <SAMPLE_ATTRIBUTE>\n"
            + "        <TAG>tag</TAG>\n"
            + "        <VALUE>International hapmap project - ceph utah residents with ancestry from northern and western europe</VALUE>\n"
            + "      </SAMPLE_ATTRIBUTE>\n"
            + "    </SAMPLE_ATTRIBUTES>\n"
            + "  </SAMPLE>\n"
            + "</SAMPLE_SET>\n";

    assertJson(
        "{\n"
            + "  \"alias\": \"GM12096\",\n"
            + "  \"accession\": \"SRS001264\",\n"
            + "  \"identifiers\": {\n"
            + "    \"externalAccessions\": [\n"
            + "      {\n"
            + "        \"db\": \"BioSample\",\n"
            + "        \"id\": \"SAMN00801146\"\n"
            + "      },\n"
            + "      {\n"
            + "        \"db\": \"Coriell\",\n"
            + "        \"id\": \"GM12096\"\n"
            + "      },\n"
            + "      {\n"
            + "        \"db\": \"Coriell\",\n"
            + "        \"id\": \"NA12096\"\n"
            + "      }\n"
            + "    ],\n"
            + "    \"secondaryAccessions\": [\n"
            + "      \"SRS001264\"\n"
            + "    ]\n"
            + "  },\n"
            + "  \"centerName\": \"Coriell\",\n"
            + "  \"title\": \"GM12096 male unaffected Ceph/utah pedigree 1332 sample\",\n"
            + "  \"description\": \"Paternal Grandfather\",\n"
            + "  \"organism\": {\n"
            + "    \"taxonId\": \"9606\",\n"
            + "    \"scientificName\": \"Homo sapiens\"\n"
            + "  },\n"
            + "  \"attributes\": [\n"
            + "    {\n"
            + "      \"tag\": \"tag\",\n"
            + "      \"value\": \"Ceph/utah pedigree 1332\"\n"
            + "    },\n"
            + "    {\n"
            + "      \"tag\": \"tag\",\n"
            + "      \"value\": \"International hapmap project - ceph utah residents with ancestry from northern and western europe\"\n"
            + "    }\n"
            + "  ],\n"
            + "  \"links\": [\n"
            + "    {\n"
            + "      \"xrefLink\": {\n"
            + "        \"db\": \"pubmed\",\n"
            + "        \"id\": \"9039096\"\n"
            + "      }\n"
            + "    },\n"
            + "    {\n"
            + "      \"xrefLink\": {\n"
            + "        \"db\": \"pubmed\",\n"
            + "        \"id\": \"18451141\"\n"
            + "      }\n"
            + "    },\n"
            + "    {\n"
            + "      \"xrefLink\": {\n"
            + "        \"db\": \"pubmed\",\n"
            + "        \"id\": \"20497546\"\n"
            + "      }\n"
            + "    }\n"
            + "  ]\n"
            + "}",
        enaSraConverter.convertSampleXmlToJson(sampleXml));
  }

  // PROJECT data model tests

  @Test
  public void testConvertSequencingProjectJsonToXml() {
    String json =
        "{\n"
            + "               \"name\":\"Test_Name\",\n"
            + "               \"title\":\"Test_Title\",\n"
            + "               \"description\":\"Test_Description\",\n"
            + "               \"sequencingProject\":{},\n"
            + "               \"links\":[\n"
            + "                  {\n"
            + "                     \"xrefLink\":{\n"
            + "                        \"db\":\"PUBMED\",\n"
            + "                        \"id\":22820990\n"
            + "                     }\n"
            + "                  }\n"
            + "               ],\n"
            + "               \"attributes\":[\n"
            + "                  {\n"
            + "                     \"tag\":\"ENA-SUBMISSION-TOOL\",\n"
            + "                     \"value\":\"SRA-Webin\"\n"
            + "                  }\n"
            + "               ]\n"
            + "            }";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<PROJECT>\n"
            + "  <NAME>Test_Name</NAME>\n"
            + "  <TITLE>Test_Title</TITLE>\n"
            + "  <DESCRIPTION>Test_Description</DESCRIPTION>\n"
            + "  <SUBMISSION_PROJECT>\n"
            + "    <SEQUENCING_PROJECT/>\n"
            + "  </SUBMISSION_PROJECT>\n"
            + "  <PROJECT_ATTRIBUTES>\n"
            + "    <PROJECT_ATTRIBUTE>\n"
            + "      <TAG>ENA-SUBMISSION-TOOL</TAG>\n"
            + "      <VALUE>SRA-Webin</VALUE>\n"
            + "    </PROJECT_ATTRIBUTE>\n"
            + "  </PROJECT_ATTRIBUTES>\n"
            + "  <PROJECT_LINKS>\n"
            + "    <PROJECT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>PUBMED</DB>\n"
            + "        <ID>22820990</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </PROJECT_LINK>\n"
            + "  </PROJECT_LINKS>\n"
            + "</PROJECT>",
        enaSraConverter.convertProjectJsonToXml(json));
  }

  @Test
  public void testConvertUmbrellaProjectJsonToXml() {
    String json =
        " {\n"
            + "                \"name\":\"Test_Name\",\n"
            + "                \"title\":\"Test_Title\",\n"
            + "                \"description\":\"Test_Description\",\n"
            + "                \"adminProject\":{\n"
            + "                   \"organism\":{\n"
            + "                      \"taxonId\":3702,\n"
            + "                      \"scientificName\":\"Arabidopsis thaliana\",\n"
            + "                      \"commonName\":\"thale cress\",\n"
            + "                      \"strain\":\"Columbia-0\"\n"
            + "                   }\n"
            + "                },\n"
            + "                \"links\":[\n"
            + "                   {\n"
            + "                      \"xrefLink\":{\n"
            + "                         \"db\":\"PUBMED\",\n"
            + "                         \"id\":22820990\n"
            + "                      }\n"
            + "                   }\n"
            + "                ],\n"
            + "                \"attributes\":[\n"
            + "                   {\n"
            + "                      \"tag\":\"ENA-SUBMISSION-TOOL\",\n"
            + "                      \"value\":\"SRA-Webin\"\n"
            + "                   }\n"
            + "                ]\n"
            + "             }";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<PROJECT>\n"
            + "  <NAME>Test_Name</NAME>\n"
            + "  <TITLE>Test_Title</TITLE>\n"
            + "  <DESCRIPTION>Test_Description</DESCRIPTION>\n"
            + "  <UMBRELLA_PROJECT/>\n"
            + "  <PROJECT_ATTRIBUTES>\n"
            + "    <PROJECT_ATTRIBUTE>\n"
            + "      <TAG>ENA-SUBMISSION-TOOL</TAG>\n"
            + "      <VALUE>SRA-Webin</VALUE>\n"
            + "    </PROJECT_ATTRIBUTE>\n"
            + "  </PROJECT_ATTRIBUTES>\n"
            + "  <PROJECT_LINKS>\n"
            + "    <PROJECT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>PUBMED</DB>\n"
            + "        <ID>22820990</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </PROJECT_LINK>\n"
            + "  </PROJECT_LINKS>\n"
            + "</PROJECT>\n",
        enaSraConverter.convertProjectJsonToXml(json));
  }

  @Test
  public void testConvertSequencingProjectJsonToXmlProjectHasRelations() {
    String json =
        "{\n"
            + "               \"alias\":\"MARC phase 1 submission project\",\n"
            + "               \"sequencingProject\":{\n"
            + "                  \"locusTagPrefixes\":[\n"
            + "                     \"BN2790\"\n"
            + "                  ]\n"
            + "               },\n"
            + "               \"identifiers\":{\n"
            + "                  \"secondaryAccessions\":[\n"
            + "                     \"ERP012322\"\n"
            + "                  ]\n"
            + "               },\n"
            + "               \"centerName\":\"European Bioinformatics Institute\",\n"
            + "               \"title\":\"MARC phase 1\",\n"
            + "               \"description\":\"A collection of ONT MinION run data submitted within MARC phase 1. The project is intended to\\n            evaluate reproducibility of performance of the MinION carried out across multiple sites. There is 20 runs\\n            from 5 laboratories in total split into two batches: 1a and 1b.\\n        \",\n"
            + "               \"attributes\":[\n"
            + "                  {\n"
            + "                     \"tag\":\"ENA-FIRST-PUBLIC\",\n"
            + "                     \"value\":\"2015-10-07\"\n"
            + "                  },\n"
            + "                  {\n"
            + "                     \"tag\":\"ENA-LAST-UPDATE\",\n"
            + "                     \"value\":\"2021-01-08\"\n"
            + "                  }\n"
            + "               ],\n"
            + "               \"links\":[\n"
            + "                  {\n"
            + "                     \"xrefLink\":{\n"
            + "                        \"db\":\"ENA-SUBMISSION\",\n"
            + "                        \"id\":\"ERA496664\"\n"
            + "                     }\n"
            + "                  },\n"
            + "                  {\n"
            + "                     \"xrefLink\":{\n"
            + "                        \"db\":\"ENA-FASTQ-FILES\",\n"
            + "                        \"id\":\"\\n                        https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n"
            + "                     }\n"
            + "                  },\n"
            + "                  {\n"
            + "                     \"xrefLink\":{\n"
            + "                        \"db\":\"ENA-SUBMITTED-FILES\",\n"
            + "                        \"id\":\"\\n                        https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format\"\n"
            + "                     }\n"
            + "                  }\n"
            + "               ],\n"
            + "               \"relationships\":[\n"
            + "                  {\n"
            + "                     \"childProject\":{\n"
            + "                        \"accession\":\"PRJEB11007\"\n"
            + "                     }\n"
            + "                  },\n"
            + "                  {\n"
            + "                     \"parentProject\":{\n"
            + "                        \"accession\":\"PRJNA12319\"\n"
            + "                     }\n"
            + "                  }\n"
            + "               ]\n"
            + "            }";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<PROJECT alias=\"MARC phase 1 submission project\" center_name=\"European Bioinformatics Institute\">\n"
            + "  <IDENTIFIERS>\n"
            + "    <SECONDARY_ID>ERP012322</SECONDARY_ID>\n"
            + "  </IDENTIFIERS>\n"
            + "  <TITLE>MARC phase 1</TITLE>\n"
            + "  <DESCRIPTION>A collection of ONT MinION run data submitted within MARC phase 1. The project is intended to\n"
            + "            evaluate reproducibility of performance of the MinION carried out across multiple sites. There is 20 runs\n"
            + "            from 5 laboratories in total split into two batches: 1a and 1b.\n"
            + "        </DESCRIPTION>\n"
            + "  <SUBMISSION_PROJECT>\n"
            + "    <SEQUENCING_PROJECT>\n"
            + "      <LOCUS_TAG_PREFIX>BN2790</LOCUS_TAG_PREFIX>\n"
            + "    </SEQUENCING_PROJECT>\n"
            + "  </SUBMISSION_PROJECT>\n"
            + "  <RELATED_PROJECTS>\n"
            + "    <RELATED_PROJECT>\n"
            + "      <CHILD_PROJECT accession=\"PRJEB11007\"/>\n"
            + "    </RELATED_PROJECT>\n"
            + "    <RELATED_PROJECT>\n"
            + "      <PARENT_PROJECT accession=\"PRJNA12319\"/>\n"
            + "    </RELATED_PROJECT>\n"
            + "  </RELATED_PROJECTS>\n"
            + "  <PROJECT_ATTRIBUTES>\n"
            + "    <PROJECT_ATTRIBUTE>\n"
            + "      <TAG>ENA-FIRST-PUBLIC</TAG>\n"
            + "      <VALUE>2015-10-07</VALUE>\n"
            + "    </PROJECT_ATTRIBUTE>\n"
            + "    <PROJECT_ATTRIBUTE>\n"
            + "      <TAG>ENA-LAST-UPDATE</TAG>\n"
            + "      <VALUE>2021-01-08</VALUE>\n"
            + "    </PROJECT_ATTRIBUTE>\n"
            + "  </PROJECT_ATTRIBUTES>\n"
            + "  <PROJECT_LINKS>\n"
            + "    <PROJECT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMISSION</DB>\n"
            + "        <ID>ERA496664</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </PROJECT_LINK>\n"
            + "    <PROJECT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-FASTQ-FILES</DB>\n"
            + "        <ID>\n"
            + "                        https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&amp;result=read_run&amp;fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </PROJECT_LINK>\n"
            + "    <PROJECT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMITTED-FILES</DB>\n"
            + "        <ID>\n"
            + "                        https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&amp;result=read_run&amp;fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </PROJECT_LINK>\n"
            + "  </PROJECT_LINKS>\n"
            + "</PROJECT>\n",
        enaSraConverter.convertProjectJsonToXml(json));
  }

  @Test
  public void testConvertSequencingProjectXmlToJson() throws JSONException {
    String xml =
        "<PROJECT_SET>\n"
            + "              <PROJECT alias=\"ena-PROJECT-17-10-2011-16:23:08:529-1\" accession=\"PRJEB6\" center_name=\"Test\">\n"
            + "                <IDENTIFIERS>\n"
            + "                  <PRIMARY_ID>PRJEB6</PRIMARY_ID>\n"
            + "                  <SUBMITTER_ID namespace=\"Test\">Test</SUBMITTER_ID>\n"
            + "                </IDENTIFIERS>\n"
            + "                <NAME>Test_Name</NAME>\n"
            + "                <TITLE>Test_Title</TITLE>\n"
            + "                <DESCRIPTION>It has recently been shown that RNA 3&apos; end formation plays a widespread role in controlling gene expression than previously thought. In order to examine the impact of regulated 3&apos; end formation genome-wide we applied Direct RNA Sequencing by Helicos to A. thaliana. With these data we show the authentic transcriptome in unprecedented detail and how 3&apos; gene end formation impacts genome organization. We reveal extreme heterogeneity in RNA 3&apos; ends, discover previously unrecognized non-coding RNAs and propose widespread re-annotation of the genome. We explain the origin of most poly(A)+ antisense RNAs and identify cis-elements that control 3&apos; end formation in different registers. These findings are essential to understand what the genome actually encodes, how it is organized and the impact of regulated 3&apos; end formation on these processes.</DESCRIPTION>\n"
            + "                <SUBMISSION_PROJECT>\n"
            + "                  <SEQUENCING_PROJECT/>\n"
            + "                  <ORGANISM>\n"
            + "                    <TAXON_ID>3702</TAXON_ID>\n"
            + "                    <SCIENTIFIC_NAME>Arabidopsis thaliana</SCIENTIFIC_NAME>\n"
            + "                    <COMMON_NAME>thale cress</COMMON_NAME>\n"
            + "                    <STRAIN>Columbia-0</STRAIN>\n"
            + "                  </ORGANISM>\n"
            + "                </SUBMISSION_PROJECT>\n"
            + "                <PROJECT_LINKS>\n"
            + "                  <PROJECT_LINK>\n"
            + "                    <XREF_LINK>\n"
            + "                      <DB>PUBMED</DB>\n"
            + "                      <ID>22820990</ID>\n"
            + "                    </XREF_LINK>\n"
            + "                  </PROJECT_LINK>\n"
            + "                </PROJECT_LINKS>\n"
            + "                <PROJECT_ATTRIBUTES>\n"
            + "                  <PROJECT_ATTRIBUTE>\n"
            + "                    <TAG>ENA-SUBMISSION-TOOL</TAG>\n"
            + "                    <VALUE>SRA-Webin</VALUE>\n"
            + "                  </PROJECT_ATTRIBUTE>\n"
            + "                </PROJECT_ATTRIBUTES>\n"
            + "              </PROJECT>\n"
            + "            </PROJECT_SET>";

    assertJson(
        "  {\n"
            + "              \"alias\" : \"ena-PROJECT-17-10-2011-16:23:08:529-1\",\n"
            + "              \"accession\" : \"PRJEB6\",\n"
            + "              \"centerName\" : \"Test\",\n"
            + "              \"name\" : \"Test_Name\",\n"
            + "              \"title\" : \"Test_Title\",\n"
            + "              \"description\" : \"It has recently been shown that RNA 3' end formation plays a widespread role in controlling gene expression than previously thought. In order to examine the impact of regulated 3' end formation genome-wide we applied Direct RNA Sequencing by Helicos to A. thaliana. With these data we show the authentic transcriptome in unprecedented detail and how 3' gene end formation impacts genome organization. We reveal extreme heterogeneity in RNA 3' ends, discover previously unrecognized non-coding RNAs and propose widespread re-annotation of the genome. We explain the origin of most poly(A)+ antisense RNAs and identify cis-elements that control 3' end formation in different registers. These findings are essential to understand what the genome actually encodes, how it is organized and the impact of regulated 3' end formation on these processes.\",\n"
            + "              \"attributes\" : [ {\n"
            + "                \"tag\" : \"ENA-SUBMISSION-TOOL\",\n"
            + "                \"value\" : \"SRA-Webin\"\n"
            + "              } ],\n"
            + "              \"links\" : [ {\n"
            + "                \"xrefLink\" : {\n"
            + "                  \"db\" : \"PUBMED\",\n"
            + "                  \"id\" : \"22820990\"\n"
            + "                }\n"
            + "              } ],\n"
            + "              \"sequencingProject\" : { \"locusTagPrefixes\" : [ ] }\n"
            + "            }",
        enaSraConverter.convertProjectXmlToJson(xml));
  }

  @Test
  public void testConvertSequencingProjectJsonToXmlProjectHasLocusTagPrefixes() {
    String json =
        "{\n"
            + "              \"alias\" : \"ena-PROJECT-DUNDEE-17-10-2011-16:23:08:529-1\",\n"
            + "              \"accession\" : \"PRJEB6\",\n"
            + "              \"identifiers\" : { },\n"
            + "              \"centerName\" : \"DUNDEE\",\n"
            + "              \"name\" : \"Study of genome-wide alternative polyadenylation in A. thaliana\",\n"
            + "              \"title\" : \"Study of genome-wide alternative polyadenylation in A. thaliana\",\n"
            + "              \"description\" : \"It has recently been shown that RNA 3' end formation plays a widespread role in controlling gene expression than previously thought. In order to examine the impact of regulated 3' end formation genome-wide we applied Direct RNA Sequencing by Helicos to A. thaliana. With these data we show the authentic transcriptome in unprecedented detail and how 3' gene end formation impacts genome organization. We reveal extreme heterogeneity in RNA 3' ends, discover previously unrecognized non-coding RNAs and propose widespread re-annotation of the genome. We explain the origin of most poly(A)+ antisense RNAs and identify cis-elements that control 3' end formation in different registers. These findings are essential to understand what the genome actually encodes, how it is organized and the impact of regulated 3' end formation on these processes.\",\n"
            + "         \n"
            + "                \"sequencingProject\" : {\n"
            + "                  \"locusTagPrefixes\" : [ \"BN2790\", \"BN2791\", \"BN2792\" ]\n"
            + "                            },\n"
            + "              \"attributes\" : [ {\n"
            + "                \"tag\" : \"ENA-SUBMISSION-TOOL\",\n"
            + "                \"value\" : \"SRA-Webin\"\n"
            + "              } ],\n"
            + "              \"links\" : [ {\n"
            + "                \"xrefLink\" : {\n"
            + "                  \"db\" : \"PUBMED\",\n"
            + "                  \"id\" : \"22820990\"\n"
            + "                }\n"
            + "              } ]\n"
            + "            }\n"
            + "            \"";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<PROJECT alias=\"ena-PROJECT-DUNDEE-17-10-2011-16:23:08:529-1\" accession=\"PRJEB6\" center_name=\"DUNDEE\">\n"
            + "  <IDENTIFIERS/>\n"
            + "  <NAME>Study of genome-wide alternative polyadenylation in A. thaliana</NAME>\n"
            + "  <TITLE>Study of genome-wide alternative polyadenylation in A. thaliana</TITLE>\n"
            + "  <DESCRIPTION>It has recently been shown that RNA 3' end formation plays a widespread role in controlling gene expression than previously thought. In order to examine the impact of regulated 3' end formation genome-wide we applied Direct RNA Sequencing by Helicos to A. thaliana. With these data we show the authentic transcriptome in unprecedented detail and how 3' gene end formation impacts genome organization. We reveal extreme heterogeneity in RNA 3' ends, discover previously unrecognized non-coding RNAs and propose widespread re-annotation of the genome. We explain the origin of most poly(A)+ antisense RNAs and identify cis-elements that control 3' end formation in different registers. These findings are essential to understand what the genome actually encodes, how it is organized and the impact of regulated 3' end formation on these processes.</DESCRIPTION>\n"
            + "  <SUBMISSION_PROJECT>\n"
            + "    <SEQUENCING_PROJECT>\n"
            + "      <LOCUS_TAG_PREFIX>BN2790</LOCUS_TAG_PREFIX>\n"
            + "      <LOCUS_TAG_PREFIX>BN2791</LOCUS_TAG_PREFIX>\n"
            + "      <LOCUS_TAG_PREFIX>BN2792</LOCUS_TAG_PREFIX>\n"
            + "    </SEQUENCING_PROJECT>\n"
            + "  </SUBMISSION_PROJECT>\n"
            + "  <PROJECT_ATTRIBUTES>\n"
            + "    <PROJECT_ATTRIBUTE>\n"
            + "      <TAG>ENA-SUBMISSION-TOOL</TAG>\n"
            + "      <VALUE>SRA-Webin</VALUE>\n"
            + "    </PROJECT_ATTRIBUTE>\n"
            + "  </PROJECT_ATTRIBUTES>\n"
            + "  <PROJECT_LINKS>\n"
            + "    <PROJECT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>PUBMED</DB>\n"
            + "        <ID>22820990</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </PROJECT_LINK>\n"
            + "  </PROJECT_LINKS>\n"
            + "</PROJECT>\n",
        enaSraConverter.convertProjectJsonToXml(json));
  }

  @Test
  public void testConvertSequencingProjectXmlToJsonProjectHasLocusTagPrefixes()
      throws JSONException {
    String xml =
        "<PROJECT_SET>\n"
            + "              <PROJECT alias=\"ena-PROJECT-DUNDEE-17-10-2011-16:23:08:529-1\" accession=\"PRJEB6\" center_name=\"DUNDEE\">\n"
            + "                <IDENTIFIERS>\n"
            + "                  <PRIMARY_ID>PRJEB6</PRIMARY_ID>\n"
            + "                  <SUBMITTER_ID namespace=\"DUNDEE\">ena-PROJECT-DUNDEE-17-10-2011-16:23:08:529-1</SUBMITTER_ID>\n"
            + "                </IDENTIFIERS>\n"
            + "                <NAME>Study of genome-wide alternative polyadenylation in A. thaliana</NAME>\n"
            + "                <TITLE>Study of genome-wide alternative polyadenylation in A. thaliana</TITLE>\n"
            + "                <DESCRIPTION>It has recently been shown that RNA 3&apos; end formation plays a widespread role in controlling gene expression than previously thought. In order to examine the impact of regulated 3&apos; end formation genome-wide we applied Direct RNA Sequencing by Helicos to A. thaliana. With these data we show the authentic transcriptome in unprecedented detail and how 3&apos; gene end formation impacts genome organization. We reveal extreme heterogeneity in RNA 3&apos; ends, discover previously unrecognized non-coding RNAs and propose widespread re-annotation of the genome. We explain the origin of most poly(A)+ antisense RNAs and identify cis-elements that control 3&apos; end formation in different registers. These findings are essential to understand what the genome actually encodes, how it is organized and the impact of regulated 3&apos; end formation on these processes.</DESCRIPTION>\n"
            + "                <SUBMISSION_PROJECT>\n"
            + "                  <SEQUENCING_PROJECT>\n"
            + "                    <LOCUS_TAG_PREFIX>BN2790</LOCUS_TAG_PREFIX>\n"
            + "                    <LOCUS_TAG_PREFIX>BN2791</LOCUS_TAG_PREFIX>\n"
            + "                    <LOCUS_TAG_PREFIX>BN2792</LOCUS_TAG_PREFIX>\n"
            + "                  </SEQUENCING_PROJECT>  \n"
            + "                  <ORGANISM>\n"
            + "                    <TAXON_ID>3702</TAXON_ID>\n"
            + "                    <SCIENTIFIC_NAME>Arabidopsis thaliana</SCIENTIFIC_NAME>\n"
            + "                    <COMMON_NAME>thale cress</COMMON_NAME>\n"
            + "                    <STRAIN>Columbia-0</STRAIN>\n"
            + "                  </ORGANISM>\n"
            + "                </SUBMISSION_PROJECT>\n"
            + "                <PROJECT_LINKS>\n"
            + "                  <PROJECT_LINK>\n"
            + "                    <XREF_LINK>\n"
            + "                      <DB>PUBMED</DB>\n"
            + "                      <ID>22820990</ID>\n"
            + "                    </XREF_LINK>\n"
            + "                  </PROJECT_LINK>\n"
            + "                </PROJECT_LINKS>\n"
            + "                <PROJECT_ATTRIBUTES>\n"
            + "                  <PROJECT_ATTRIBUTE>\n"
            + "                    <TAG>ENA-SUBMISSION-TOOL</TAG>\n"
            + "                    <VALUE>SRA-Webin</VALUE>\n"
            + "                  </PROJECT_ATTRIBUTE>\n"
            + "                </PROJECT_ATTRIBUTES>\n"
            + "              </PROJECT>\n"
            + "            </PROJECT_SET>";

    assertJson(
        "{\n"
            + "              \"alias\" : \"ena-PROJECT-DUNDEE-17-10-2011-16:23:08:529-1\",\n"
            + "              \"accession\" : \"PRJEB6\",\n"
            + "              \"centerName\" : \"DUNDEE\",\n"
            + "              \"name\" : \"Study of genome-wide alternative polyadenylation in A. thaliana\",\n"
            + "              \"title\" : \"Study of genome-wide alternative polyadenylation in A. thaliana\",\n"
            + "              \"description\" : \"It has recently been shown that RNA 3' end formation plays a widespread role in controlling gene expression than previously thought. In order to examine the impact of regulated 3' end formation genome-wide we applied Direct RNA Sequencing by Helicos to A. thaliana. With these data we show the authentic transcriptome in unprecedented detail and how 3' gene end formation impacts genome organization. We reveal extreme heterogeneity in RNA 3' ends, discover previously unrecognized non-coding RNAs and propose widespread re-annotation of the genome. We explain the origin of most poly(A)+ antisense RNAs and identify cis-elements that control 3' end formation in different registers. These findings are essential to understand what the genome actually encodes, how it is organized and the impact of regulated 3' end formation on these processes.\",\n"
            + "              \"attributes\" : [ {\n"
            + "                \"tag\" : \"ENA-SUBMISSION-TOOL\",\n"
            + "                \"value\" : \"SRA-Webin\"\n"
            + "              } ],\n"
            + "              \"links\" : [ {\n"
            + "                \"xrefLink\" : {\n"
            + "                  \"db\" : \"PUBMED\",\n"
            + "                  \"id\" : \"22820990\"\n"
            + "                }\n"
            + "              } ],\n"
            + "              \"sequencingProject\" : {\n"
            + "                \"locusTagPrefixes\" : [ \"BN2790\", \"BN2791\", \"BN2792\" ]\n"
            + "              }\n"
            + "            }",
        enaSraConverter.convertProjectXmlToJson(xml));
  }

  @Test
  public void testConvertSequencingProjectXmlToJsonProjectHasRelations() throws JSONException {
    String xml =
        " <PROJECT accession=\"PRJEB11008\" alias=\"MARC phase 1 submission project\"\n"
            + "                         center_name=\"European Bioinformatics Institute\">\n"
            + "                    <IDENTIFIERS>\n"
            + "                        <PRIMARY_ID>PRJEB11008</PRIMARY_ID>\n"
            + "                        <SECONDARY_ID>ERP012322</SECONDARY_ID>\n"
            + "                        <SUBMITTER_ID namespace=\"European Bioinformatics Institute\">MARC phase 1 submission project</SUBMITTER_ID>\n"
            + "                    </IDENTIFIERS>\n"
            + "                    <TITLE>MARC phase 1</TITLE>\n"
            + "                    <DESCRIPTION>A collection of ONT MinION run data submitted within MARC phase 1. The project is intended to\n"
            + "                        evaluate reproducibility of performance of the MinION carried out across multiple sites. There is 20 runs\n"
            + "                        from 5 laboratories in total split into two batches: 1a and 1b.\n"
            + "                    </DESCRIPTION>\n"
            + "                    <SUBMISSION_PROJECT>\n"
            + "                        <SEQUENCING_PROJECT>\n"
            + "                            <LOCUS_TAG_PREFIX>BN2790</LOCUS_TAG_PREFIX>\n"
            + "                        </SEQUENCING_PROJECT>\n"
            + "                        <ORGANISM>\n"
            + "                            <TAXON_ID>511145</TAXON_ID>\n"
            + "                            <SCIENTIFIC_NAME>Escherichia coli str. K-12 substr. MG1655</SCIENTIFIC_NAME>\n"
            + "                        </ORGANISM>\n"
            + "                    </SUBMISSION_PROJECT>\n"
            + "                    <RELATED_PROJECTS>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB11007\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9084\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9088\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9111\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9131\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9135\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9138\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9289\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9352\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9442\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9447\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9448\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9455\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <CHILD_PROJECT accession=\"PRJEB9456\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                        <RELATED_PROJECT>\n"
            + "                            <PARENT_PROJECT accession=\"PRJNA12319\"/>\n"
            + "                        </RELATED_PROJECT>\n"
            + "                    </RELATED_PROJECTS>\n"
            + "                    <PROJECT_LINKS>\n"
            + "                        <PROJECT_LINK>\n"
            + "                            <XREF_LINK>\n"
            + "                                <DB>ENA-SUBMISSION</DB>\n"
            + "                                <ID>ERA496664</ID>\n"
            + "                            </XREF_LINK>\n"
            + "                        </PROJECT_LINK>\n"
            + "                        <PROJECT_LINK>\n"
            + "                            <XREF_LINK>\n"
            + "                                <DB>ENA-FASTQ-FILES</DB>\n"
            + "                                <ID>\n"
            + "                                    <![CDATA[https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes]]></ID>\n"
            + "                            </XREF_LINK>\n"
            + "                        </PROJECT_LINK>\n"
            + "                        <PROJECT_LINK>\n"
            + "                            <XREF_LINK>\n"
            + "                                <DB>ENA-SUBMITTED-FILES</DB>\n"
            + "                                <ID>\n"
            + "                                    <![CDATA[https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format]]></ID>\n"
            + "                            </XREF_LINK>\n"
            + "                        </PROJECT_LINK>\n"
            + "                    </PROJECT_LINKS>\n"
            + "                    <PROJECT_ATTRIBUTES>\n"
            + "                        <PROJECT_ATTRIBUTE>\n"
            + "                            <TAG>ENA-FIRST-PUBLIC</TAG>\n"
            + "                            <VALUE>2015-10-07</VALUE>\n"
            + "                        </PROJECT_ATTRIBUTE>\n"
            + "                        <PROJECT_ATTRIBUTE>\n"
            + "                            <TAG>ENA-LAST-UPDATE</TAG>\n"
            + "                            <VALUE>2021-01-08</VALUE>\n"
            + "                        </PROJECT_ATTRIBUTE>\n"
            + "                    </PROJECT_ATTRIBUTES>\n"
            + "                </PROJECT>";

    assertJson(
        "{\n"
            + "  \"alias\" : \"MARC phase 1 submission project\",\n"
            + "  \"accession\" : \"PRJEB11008\",\n"
            + "  \"identifiers\" : {\n"
            + "    \"secondaryAccessions\" : [ \"ERP012322\" ]\n"
            + "  },\n"
            + "  \"centerName\" : \"European Bioinformatics Institute\",\n"
            + "  \"title\" : \"MARC phase 1\",\n"
            + "  \"description\" : \"A collection of ONT MinION run data submitted within MARC phase 1. The project is intended to\\n                        evaluate reproducibility of performance of the MinION carried out across multiple sites. There is 20 runs\\n                        from 5 laboratories in total split into two batches: 1a and 1b.\\n                    \",\n"
            + "  \"sequencingProject\" : {\n"
            + "    \"locusTagPrefixes\" : [ \"BN2790\" ]\n"
            + "  },\n"
            + "  \"relationships\" : [ {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB11007\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9084\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9088\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9111\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9131\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9135\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9138\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9289\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9352\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9442\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9447\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9448\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9455\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"childProject\" : {\n"
            + "      \"accession\" : \"PRJEB9456\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"parentProject\" : {\n"
            + "      \"accession\" : \"PRJNA12319\"\n"
            + "    }\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"ENA-FIRST-PUBLIC\",\n"
            + "    \"value\" : \"2015-10-07\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"ENA-LAST-UPDATE\",\n"
            + "    \"value\" : \"2021-01-08\"\n"
            + "  } ],\n"
            + "  \"links\" : [ {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SUBMISSION\",\n"
            + "      \"id\" : \"ERA496664\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-FASTQ-FILES\",\n"
            + "      \"id\" : \"\\n                                    https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SUBMITTED-FILES\",\n"
            + "      \"id\" : \"\\n                                    https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB11008&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format\"\n"
            + "    }\n"
            + "  } ]\n"
            + "}\n",
        enaSraConverter.convertProjectXmlToJson(xml));
  }

  @Test
  public void testExperimentJsonConversionToXmlExperimentLibraryLayoutPaired() {
    String json =
        "{\n"
            + "   \"alias\":\"ena-EXPERIMENT-Graz University of Technical-11-01-2021-14:17:33:319-5\",\n"
            + "   \"accession\":\"ERX4866417\",\n"
            + "   \"centerName\":\"Graz University of Technical\",\n"
            + "   \"title\":\"Illumina MiSeq sequencing\",\n"
            + "   \"study\":{\n"
            + "      \"accession\":\"ERP121346\"\n"
            + "   },\n"
            + "   \"samples\":[\n"
            + "      {\n"
            + "         \"accession\":\"ERS121346\"\n"
            + "      },\n"
            + "      {\n"
            + "         \"accession\":\"ERS121345\"\n"
            + "      },\n"
            + "      {\n"
            + "         \"alias\":\"sample1\"\n"
            + "      }\n"
            + "   ],\n"
            + "   \"designDescription\":\"\",\n"
            + "   \"libraryDescriptor\":{\n"
            + "      \"libraryName\":\"unspecified\",\n"
            + "      \"libraryStrategy\":\"AMPLICON\",\n"
            + "      \"librarySource\":\"GENOMIC\",\n"
            + "      \"librarySelection\":\"PCR\",\n"
            + "      \"libraryLayout\":\"paired\"\n"
            + "   },\n"
            + "   \"instrumentPlatform\":\"ILLUMINA\",\n"
            + "   \"instrumentModel\":\"Illumina MiSeq\",\n"
            + "   \"attributes\":[\n"
            + "      {\n"
            + "         \"tag\":\"test_tag\",\n"
            + "         \"value\":\"test_val\"\n"
            + "      },\n"
            + "      {\n"
            + "         \"tag\":\"test_tag_1\",\n"
            + "         \"value\":\"test_val_1\"\n"
            + "      }\n"
            + "   ],\n"
            + "   \"links\":[\n"
            + "      {\n"
            + "         \"xrefLink\":{\n"
            + "            \"db\":\"ENA-SAMPLE\",\n"
            + "            \"id\":\"ERS5524413\"\n"
            + "         }\n"
            + "      },\n"
            + "      {\n"
            + "         \"xrefLink\":{\n"
            + "            \"db\":\"ENA-RUN\",\n"
            + "            \"id\":\"ERR5060348\"\n"
            + "         }\n"
            + "      },\n"
            + "      {\n"
            + "         \"xrefLink\":{\n"
            + "            \"db\":\"ENA-SUBMISSION\",\n"
            + "            \"id\":\"ERA3206930\"\n"
            + "         }\n"
            + "      },\n"
            + "      {\n"
            + "         \"xrefLink\":{\n"
            + "            \"db\":\"ENA-FASTQ-FILES\",\n"
            + "            \"id\":\"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n"
            + "         }\n"
            + "      },\n"
            + "      {\n"
            + "         \"xrefLink\":{\n"
            + "            \"db\":\"ENA-SUBMITTED-FILES\",\n"
            + "            \"id\":\"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format\"\n"
            + "         }\n"
            + "      }\n"
            + "   ]\n"
            + "}";

    assertXml(
        enaSraConverter.convertExperimentJsonToXml(json),
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<EXPERIMENT alias=\"ena-EXPERIMENT-Graz University of Technical-11-01-2021-14:17:33:319-5\" accession=\"ERX4866417\" center_name=\"Graz University of Technical\">\n"
            + "  <TITLE>Illumina MiSeq sequencing</TITLE>\n"
            + "  <STUDY_REF accession=\"ERP121346\"/>\n"
            + "  <DESIGN>\n"
            + "    <DESIGN_DESCRIPTION></DESIGN_DESCRIPTION>\n"
            + "    <SAMPLE_DESCRIPTOR accession=\"ERS121346\"/>\n"
            + "    <SAMPLE_DESCRIPTOR accession=\"ERS121345\"/>\n"
            + "    <SAMPLE_DESCRIPTOR refname=\"sample1\"/>\n"
            + "    <LIBRARY_DESCRIPTOR>\n"
            + "      <LIBRARY_NAME>unspecified</LIBRARY_NAME>\n"
            + "      <LIBRARY_STRATEGY>AMPLICON</LIBRARY_STRATEGY>\n"
            + "      <LIBRARY_SOURCE>GENOMIC</LIBRARY_SOURCE>\n"
            + "      <LIBRARY_SELECTION>PCR</LIBRARY_SELECTION>\n"
            + "      <LIBRARY_LAYOUT>\n"
            + "        <PAIRED/>\n"
            + "      </LIBRARY_LAYOUT>\n"
            + "    </LIBRARY_DESCRIPTOR>\n"
            + "  </DESIGN>\n"
            + "  <PLATFORM>\n"
            + "    <ILLUMINA>\n"
            + "      <INSTRUMENT_MODEL>Illumina MiSeq</INSTRUMENT_MODEL>\n"
            + "    </ILLUMINA>\n"
            + "  </PLATFORM>\n"
            + "  <EXPERIMENT_ATTRIBUTES>\n"
            + "    <EXPERIMENT_ATTRIBUTE>\n"
            + "      <TAG>test_tag</TAG>\n"
            + "      <VALUE>test_val</VALUE>\n"
            + "    </EXPERIMENT_ATTRIBUTE>\n"
            + "    <EXPERIMENT_ATTRIBUTE>\n"
            + "      <TAG>test_tag_1</TAG>\n"
            + "      <VALUE>test_val_1</VALUE>\n"
            + "    </EXPERIMENT_ATTRIBUTE>\n"
            + "  </EXPERIMENT_ATTRIBUTES>\n"
            + "  <EXPERIMENT_LINKS>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SAMPLE</DB>\n"
            + "        <ID>ERS5524413</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-RUN</DB>\n"
            + "        <ID>ERR5060348</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMISSION</DB>\n"
            + "        <ID>ERA3206930</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-FASTQ-FILES</DB>\n"
            + "        <ID>https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&amp;result=read_run&amp;fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMITTED-FILES</DB>\n"
            + "        <ID>https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&amp;result=read_run&amp;fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "  </EXPERIMENT_LINKS>\n"
            + "</EXPERIMENT>\n");
  }

  @Test
  public void testExperimentXmlConversionToJson() throws JSONException {
    String xml =
        "<EXPERIMENT alias=\"ena-EXPERIMENT-Graz University of Technical-11-01-2021-14:17:33:319-5\" accession=\"ERX4866417\" center_name=\"Graz University of Technical\">\n"
            + "  <TITLE>Illumina MiSeq sequencing</TITLE>\n"
            + "  <STUDY_REF accession=\"ERP121346\"/>\n"
            + "  <DESIGN>\n"
            + "      <SAMPLE_DESCRIPTOR accession=\"ERS121346\"/>\n"
            + "      <SAMPLE_DESCRIPTOR accession=\"ERS121345\"/>\n"
            + "      <SAMPLE_DESCRIPTOR refname=\"sample1\"/>\n"
            + "    <LIBRARY_DESCRIPTOR>\n"
            + "      <LIBRARY_NAME>unspecified</LIBRARY_NAME>\n"
            + "      <LIBRARY_STRATEGY>AMPLICON</LIBRARY_STRATEGY>\n"
            + "      <LIBRARY_SOURCE>GENOMIC</LIBRARY_SOURCE>\n"
            + "      <LIBRARY_SELECTION>PCR</LIBRARY_SELECTION>\n"
            + "      <LIBRARY_LAYOUT>\n"
            + "        <PAIRED/>\n"
            + "      </LIBRARY_LAYOUT>\n"
            + "    </LIBRARY_DESCRIPTOR>\n"
            + "  </DESIGN>\n"
            + "  <PLATFORM>\n"
            + "    <ILLUMINA>\n"
            + "      <INSTRUMENT_MODEL>Illumina MiSeq</INSTRUMENT_MODEL>\n"
            + "    </ILLUMINA>\n"
            + "  </PLATFORM>\n"
            + "  <EXPERIMENT_ATTRIBUTES>\n"
            + "    <EXPERIMENT_ATTRIBUTE>\n"
            + "      <TAG>test_tag</TAG>\n"
            + "      <VALUE>test_val</VALUE>\n"
            + "    </EXPERIMENT_ATTRIBUTE>\n"
            + "    <EXPERIMENT_ATTRIBUTE>\n"
            + "      <TAG>test_tag_1</TAG>\n"
            + "      <VALUE>test_val_1</VALUE>\n"
            + "    </EXPERIMENT_ATTRIBUTE>\n"
            + "  </EXPERIMENT_ATTRIBUTES>\n"
            + "  <EXPERIMENT_LINKS>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SAMPLE</DB>\n"
            + "        <ID>ERS5524413</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-RUN</DB>\n"
            + "        <ID>ERR5060348</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMISSION</DB>\n"
            + "        <ID>ERA3206930</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-FASTQ-FILES</DB>\n"
            + "        <ID>https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&amp;result=read_run&amp;fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "    <EXPERIMENT_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMITTED-FILES</DB>\n"
            + "        <ID>https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&amp;result=read_run&amp;fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </EXPERIMENT_LINK>\n"
            + "  </EXPERIMENT_LINKS>\n"
            + "</EXPERIMENT>";

    assertJson(
        enaSraConverter.convertExperimentXmlToJson(xml),
        "{\n"
            + "  \"alias\" : \"ena-EXPERIMENT-Graz University of Technical-11-01-2021-14:17:33:319-5\",\n"
            + "  \"accession\" : \"ERX4866417\",\n"
            + "  \"identifiers\" : null,\n"
            + "  \"centerName\" : \"Graz University of Technical\",\n"
            + "  \"title\" : \"Illumina MiSeq sequencing\",\n"
            + "  \"study\" : {\n"
            + "    \"accession\" : \"ERP121346\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"accession\" : \"ERS121346\"\n"
            + "  }, {\n"
            + "    \"accession\" : \"ERS121345\"\n"
            + "  }, {\n"
            + "    \"alias\" : \"sample1\"\n"
            + "  } ],\n"
            + "  \"designDescription\" : \"\",\n"
            + "  \"libraryDescriptor\" : {\n"
            + "    \"libraryName\" : \"unspecified\",\n"
            + "    \"libraryStrategy\" : \"AMPLICON\",\n"
            + "    \"librarySource\" : \"GENOMIC\",\n"
            + "    \"librarySelection\" : \"PCR\",\n"
            + "    \"libraryLayout\" : \"PAIRED\"\n"
            + "  },\n"
            + "  \"instrumentPlatform\" : \"ILLUMINA\",\n"
            + "  \"instrumentModel\" : \"Illumina MiSeq\",\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"test_tag\",\n"
            + "    \"value\" : \"test_val\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"test_tag_1\",\n"
            + "    \"value\" : \"test_val_1\"\n"
            + "  } ],\n"
            + "  \"links\" : [ {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SAMPLE\",\n"
            + "      \"id\" : \"ERS5524413\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-RUN\",\n"
            + "      \"id\" : \"ERR5060348\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SUBMISSION\",\n"
            + "      \"id\" : \"ERA3206930\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-FASTQ-FILES\",\n"
            + "      \"id\" : \"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SUBMITTED-FILES\",\n"
            + "      \"id\" : \"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format\"\n"
            + "    }\n"
            + "  } ]\n"
            + "}");
  }

  @Test
  public void testRunJsonConversionToXml() {
    String json =
        "{\n"
            + "                       \"alias\":\"run_alias_1\",\n"
            + "                       \"accession\":\"ERR12345\",\n"
            + "                       \"identifiers\":null,\n"
            + "                       \"centerName\":\"test\",\n"
            + "                       \"brokerName\":\"test\",\n"
            + "                       \"title\":\"test\",\n"
            + "                       \"description\":\"test\",\n"
            + "                       \"experiment\":{\n"
            + "                          \"accession\":\"ERX12345\"\n"
            + "                       },\n"
            + "                       \"instrumentPlatform\":\"ILLUMINA\",\n"
            + "                       \"instrumentModel\":\"Illumina MiSeq\",\n"
            + "                       \"files\":[\n"
            + "                          {\n"
            + "                             \"readType\" : [ \"single\", \"paired\" ],\n"
            + "                             \"fileName\":\"24_rice_lines/B231/file1.fq.gz\",\n"
            + "                             \"fileType\":\"fastq\",\n"
            + "                             \"checksumMethod\":\"MD5\",\n"
            + "                             \"checksum\":\"2ba69754fee276fd62d7310068d3e07d\"\n"
            + "                          },\n"
            + "                          {\n"
            + "                             \"readType\":[ \"single\"],\n"
            + "                             \"fileName\":\"file1.fq.gz\",\n"
            + "                             \"fileType\":\"fastq\",\n"
            + "                             \"checksumMethod\":\"MD5\",\n"
            + "                             \"checksum\":\"03574ac396b01a04cf3143167c7abd89\"\n"
            + "                          }\n"
            + "                       ],\n"
            + "                       \"attributes\":[\n"
            + "                          {\n"
            + "                             \"tag\":\"test_tag\",\n"
            + "                             \"value\":\"test_val\"\n"
            + "                          },\n"
            + "                          {\n"
            + "                             \"tag\":\"test_tag_1\",\n"
            + "                             \"value\":\"test_val_1\"\n"
            + "                          }\n"
            + "                       ],\n"
            + "                       \"links\":[\n"
            + "                          {\n"
            + "                             \"xrefLink\":{\n"
            + "                                \"db\":\"ENA-SAMPLE\",\n"
            + "                                \"id\":\"ERS5524413\"\n"
            + "                             }\n"
            + "                          },\n"
            + "                          {\n"
            + "                             \"xrefLink\":{\n"
            + "                                \"db\":\"ENA-RUN\",\n"
            + "                                \"id\":\"ERR5060348\"\n"
            + "                             }\n"
            + "                          },\n"
            + "                          {\n"
            + "                             \"xrefLink\":{\n"
            + "                                \"db\":\"ENA-SUBMISSION\",\n"
            + "                                \"id\":\"ERA3206930\"\n"
            + "                             }\n"
            + "                          },\n"
            + "                          {\n"
            + "                             \"xrefLink\":{\n"
            + "                                \"db\":\"ENA-FASTQ-FILES\",\n"
            + "                                \"id\":\"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n"
            + "                             }\n"
            + "                          },\n"
            + "                          {\n"
            + "                             \"xrefLink\":{\n"
            + "                                \"db\":\"ENA-SUBMITTED-FILES\",\n"
            + "                                \"id\":\"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format\"\n"
            + "                             }\n"
            + "                          }\n"
            + "                       ]\n"
            + "                    }";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<RUN alias=\"run_alias_1\" accession=\"ERR12345\" center_name=\"test\" broker_name=\"test\">\n"
            + "  <TITLE>test</TITLE>\n"
            + "  <EXPERIMENT_REF accession=\"ERX12345\"/>\n"
            + "  <PLATFORM>\n"
            + "    <ILLUMINA>\n"
            + "      <INSTRUMENT_MODEL>Illumina MiSeq</INSTRUMENT_MODEL>\n"
            + "    </ILLUMINA>\n"
            + "  </PLATFORM>\n"
            + "  <DATA_BLOCK>\n"
            + "    <FILES>\n"
            + "      <FILE filename=\"24_rice_lines/B231/file1.fq.gz\" filetype=\"fastq\" checksum_method=\"MD5\" checksum=\"2ba69754fee276fd62d7310068d3e07d\">\n"
            + "        <READ_TYPE>single</READ_TYPE>\n"
            + "        <READ_TYPE>paired</READ_TYPE>\n"
            + "      </FILE>\n"
            + "      <FILE filename=\"file1.fq.gz\" filetype=\"fastq\" checksum_method=\"MD5\" checksum=\"03574ac396b01a04cf3143167c7abd89\">\n"
            + "        <READ_TYPE>single</READ_TYPE>\n"
            + "      </FILE>\n"
            + "    </FILES>\n"
            + "  </DATA_BLOCK>\n"
            + "  <RUN_ATTRIBUTES>\n"
            + "    <RUN_ATTRIBUTE>\n"
            + "      <TAG>test_tag</TAG>\n"
            + "      <VALUE>test_val</VALUE>\n"
            + "    </RUN_ATTRIBUTE>\n"
            + "    <RUN_ATTRIBUTE>\n"
            + "      <TAG>test_tag_1</TAG>\n"
            + "      <VALUE>test_val_1</VALUE>\n"
            + "    </RUN_ATTRIBUTE>\n"
            + "  </RUN_ATTRIBUTES>\n"
            + "  <RUN_LINKS>\n"
            + "    <RUN_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SAMPLE</DB>\n"
            + "        <ID>ERS5524413</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </RUN_LINK>\n"
            + "    <RUN_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-RUN</DB>\n"
            + "        <ID>ERR5060348</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </RUN_LINK>\n"
            + "    <RUN_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMISSION</DB>\n"
            + "        <ID>ERA3206930</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </RUN_LINK>\n"
            + "    <RUN_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-FASTQ-FILES</DB>\n"
            + "        <ID>https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&amp;result=read_run&amp;fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </RUN_LINK>\n"
            + "    <RUN_LINK>\n"
            + "      <XREF_LINK>\n"
            + "        <DB>ENA-SUBMITTED-FILES</DB>\n"
            + "        <ID>https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERX4866417&amp;result=read_run&amp;fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format</ID>\n"
            + "      </XREF_LINK>\n"
            + "    </RUN_LINK>\n"
            + "  </RUN_LINKS>\n"
            + "</RUN>\n",
        enaSraConverter.convertRunJsonToXml(json));
  }

  @Test
  public void testRunXmlConversionToJson() throws JSONException {
    String xml =
        "<RUN_SET>\n"
            + "                        <RUN accession=\"ERR5060344\" alias=\"ena-RUN-Graz University of Technical-11-01-2021-14:17:33:318-1\" center_name=\"Graz University of Technical\">\n"
            + "                            <IDENTIFIERS>\n"
            + "                                <PRIMARY_ID>ERR5060344</PRIMARY_ID>\n"
            + "                                <SUBMITTER_ID namespace=\"Graz University of Technical\">ena-RUN-Graz University of Technical-11-01-2021-14:17:33:318-1</SUBMITTER_ID>\n"
            + "                            </IDENTIFIERS>\n"
            + "                            <TITLE>Illumina MiSeq sequencing</TITLE>\n"
            + "                            <EXPERIMENT_REF accession=\"ERX4866413\">\n"
            + "                                <IDENTIFIERS>\n"
            + "                                    <PRIMARY_ID>ERX4866413</PRIMARY_ID>\n"
            + "                                </IDENTIFIERS>\n"
            + "                            </EXPERIMENT_REF>\n"
            + "                            <PLATFORM>\n"
            + "                                <ILLUMINA>\n"
            + "                                    <INSTRUMENT_MODEL>Illumina MiSeq</INSTRUMENT_MODEL>\n"
            + "                                </ILLUMINA>\n"
            + "                            </PLATFORM>\n"
            + "                            <DATA_BLOCK>\n"
            + "                                <FILES>\n"
            + "                                    <FILE checksum=\"66D2AF73857D6085D9ABA2B91F62B9EA\" checksum_method=\"MD5\" filename=\"run/ERR506/ERR5060344/VitisC_16S_1_8_L001_R1_001.fastq.gz\" filetype=\"fastq\" quality_scoring_system=\"phred\">\n"
            + "                                        <READ_TYPE>single</READ_TYPE>\n"
            + "                                        <READ_TYPE>paired</READ_TYPE>\n"
            + "                                     </FILE>\n"
            + "                                </FILES>\n"
            + "                            </DATA_BLOCK>\n"
            + "                            <RUN_LINKS>\n"
            + "                                <RUN_LINK>\n"
            + "                                    <XREF_LINK>\n"
            + "                                        <DB>ENA-STUDY</DB>\n"
            + "                                        <ID>ERP121345</ID>\n"
            + "                                    </XREF_LINK>\n"
            + "                                </RUN_LINK>\n"
            + "                                <RUN_LINK>\n"
            + "                                    <XREF_LINK>\n"
            + "                                        <DB>ENA-SAMPLE</DB>\n"
            + "                                        <ID>ERS5524409</ID>\n"
            + "                                    </XREF_LINK>\n"
            + "                                </RUN_LINK>\n"
            + "                                <RUN_LINK>\n"
            + "                                    <XREF_LINK>\n"
            + "                                        <DB>ENA-SUBMISSION</DB>\n"
            + "                                        <ID>ERA3206930</ID>\n"
            + "                                    </XREF_LINK>\n"
            + "                                </RUN_LINK>\n"
            + "                                <RUN_LINK>\n"
            + "                                    <XREF_LINK>\n"
            + "                                        <DB>ENA-FASTQ-FILES</DB>\n"
            + "                                        <ID><![CDATA[https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERR5060344&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes]]></ID>\n"
            + "                                    </XREF_LINK>\n"
            + "                                </RUN_LINK>\n"
            + "                                <RUN_LINK>\n"
            + "                                    <XREF_LINK>\n"
            + "                                        <DB>ENA-SUBMITTED-FILES</DB>\n"
            + "                                        <ID><![CDATA[https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERR5060344&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format]]></ID>\n"
            + "                                    </XREF_LINK>\n"
            + "                                </RUN_LINK>\n"
            + "                            </RUN_LINKS>\n"
            + "                            <RUN_ATTRIBUTES>\n"
            + "                                <RUN_ATTRIBUTE>\n"
            + "                                    <TAG>ENA-FIRST-PUBLIC</TAG>\n"
            + "                                    <VALUE>2021-01-11</VALUE>\n"
            + "                                </RUN_ATTRIBUTE>\n"
            + "                                <RUN_ATTRIBUTE>\n"
            + "                                    <TAG>ENA-LAST-UPDATE</TAG>\n"
            + "                                    <VALUE>2021-01-11</VALUE>\n"
            + "                                </RUN_ATTRIBUTE>\n"
            + "                            </RUN_ATTRIBUTES>\n"
            + "                        </RUN>\n"
            + "                    </RUN_SET>";

    assertJson(
        "{\n"
            + "  \"alias\" : \"ena-RUN-Graz University of Technical-11-01-2021-14:17:33:318-1\",\n"
            + "  \"accession\" : \"ERR5060344\",\n"
            + "  \"identifiers\" : { },\n"
            + "  \"centerName\" : \"Graz University of Technical\",\n"
            + "  \"title\" : \"Illumina MiSeq sequencing\",\n"
            + "  \"experiment\" : {\n"
            + "    \"accession\" : \"ERX4866413\"\n"
            + "  },\n"
            + "  \"instrumentPlatform\" : \"ILLUMINA\",\n"
            + "  \"instrumentModel\" : \"Illumina MiSeq\",\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"fastq\",\n"
            + "    \"checksum\" : \"66D2AF73857D6085D9ABA2B91F62B9EA\",\n"
            + "    \"readType\" : [ \"single\", \"paired\" ],\n"
            + "    \"fileName\" : \"run/ERR506/ERR5060344/VitisC_16S_1_8_L001_R1_001.fastq.gz\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"ENA-FIRST-PUBLIC\",\n"
            + "    \"value\" : \"2021-01-11\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"ENA-LAST-UPDATE\",\n"
            + "    \"value\" : \"2021-01-11\"\n"
            + "  } ],\n"
            + "  \"links\" : [ {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-STUDY\",\n"
            + "      \"id\" : \"ERP121345\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SAMPLE\",\n"
            + "      \"id\" : \"ERS5524409\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SUBMISSION\",\n"
            + "      \"id\" : \"ERA3206930\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-FASTQ-FILES\",\n"
            + "      \"id\" : \"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERR5060344&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"xrefLink\" : {\n"
            + "      \"db\" : \"ENA-SUBMITTED-FILES\",\n"
            + "      \"id\" : \"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=ERR5060344&result=read_run&fields=run_accession,submitted_ftp,submitted_md5,submitted_bytes,submitted_format\"\n"
            + "    }\n"
            + "  } ]\n"
            + "}\n",
        enaSraConverter.convertRunXmlToJson(xml));
  }

  @Test
  /*
  GENOME_MAP analysis type has a list of key-value pairs, hence testing this simple type
   */
  public void testAnalysisXml_GenomeMapAnalysisType_ConversionToJson() throws JSONException {
    String xml =
        "<?xml version = '1.0' encoding = 'UTF-8'?>\n"
            + "<ANALYSIS_SET>\n"
            + "    <ANALYSIS alias=\"\" center_name=\"\">\n"
            + "        <TITLE>Pathogen_analysis title</TITLE>\n"
            + "        <STUDY_REF refname=\"\" />\n"
            + "        <SAMPLE_REF refname=\"\" />\n"
            + "        <ANALYSIS_TYPE>\n"
            + "            <GENOME_MAP>\n"
            + "                <PROGRAM>test_genome_map_program</PROGRAM>\n"
            + "                <PLATFORM>BioNano</PLATFORM>\n"
            + "                <DESCRIPTION>test_genome_map_description</DESCRIPTION>\n"
            + "            </GENOME_MAP>\n"
            + "        </ANALYSIS_TYPE>\n"
            + "        <FILES>\n"
            + "            <FILE filename=\"sequence_flatfile.tsv.gz\" filetype=\"tab\" checksum_method=\"MD5\" checksum=\"5a8018b5d369af96ccc07dc77b923fb1\" />\n"
            + "            <FILE filename=\"test.bnx\" filetype=\"BioNano_native\" checksum_method=\"MD5\" checksum=\"16cf6f36e00ee61ba888ae14efd83873\"/>\n"
            + "            <FILE filename=\"dir/test.bnx\" filetype=\"BioNano_native\" checksum_method=\"MD5\" checksum=\"16cf6f36e00ee61ba888ae14efd83873\"/>\n"
            + "        </FILES>\n"
            + "    </ANALYSIS>\n"
            + "</ANALYSIS_SET>\n"
            + "\n";

    assertJson(
        "{\n"
            + "  \"alias\" : \"\",\n"
            + "  \"centerName\" : \"\",\n"
            + "  \"title\" : \"Pathogen_analysis title\",\n"
            + "  \"study\" : {\n"
            + "    \"alias\" : \"\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"alias\" : \"\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"GENOME_MAP\",\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"tab\",\n"
            + "    \"checksum\" : \"5a8018b5d369af96ccc07dc77b923fb1\",\n"
            + "    \"fileName\" : \"sequence_flatfile.tsv.gz\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  }, {\n"
            + "    \"fileType\" : \"BioNano_native\",\n"
            + "    \"checksum\" : \"16cf6f36e00ee61ba888ae14efd83873\",\n"
            + "    \"fileName\" : \"test.bnx\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  }, {\n"
            + "    \"fileType\" : \"BioNano_native\",\n"
            + "    \"checksum\" : \"16cf6f36e00ee61ba888ae14efd83873\",\n"
            + "    \"fileName\" : \"dir/test.bnx\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"PLATFORM\",\n"
            + "    \"value\" : \"BioNano\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"DESCRIPTION\",\n"
            + "    \"value\" : \"test_genome_map_description\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"PROGRAM\",\n"
            + "    \"value\" : \"test_genome_map_program\"\n"
            + "  } ]\n"
            + "}\n",
        enaSraConverter.convertAnalysisXmlToJson(xml));
  }

  @Test
  /*
  ASSEMBLY_GRAPH analysis type has the ASSEMBLY complex type and can have multiple occurrences, hence testing this complex type
   */
  public void testAnalysisXml_AssemblyGraphAnalysisType_ConversionToJson() throws JSONException {
    String xml =
        "<ANALYSIS_SET>\n"
            + "    <ANALYSIS alias=\"graph_X_PGGB\">\n"
            + "        <TITLE>Pangenome assembly graph X - generated using PGGB pipeline</TITLE>\n"
            + "        <DESCRIPTION>Pangenome assembly graph X was generated . Etc. </DESCRIPTION>\n"
            + "        <STUDY_REF accession=\"ERP123456\"/>\n"
            + "        <SAMPLE_REF accession=\"ERS123456\"/>\n"
            + "        <SAMPLE_REF accession=\"ERS123457\"/>\n"
            + "        <SAMPLE_REF accession=\"ERS123458\"/>\n"
            + "        <ANALYSIS_TYPE>\n"
            + "            <ASSEMBLY_GRAPH>\n"
            + "                <ASSEMBLY>\n"
            + "                    <STANDARD accession=\"GCA_012345.1\"/>\n"
            + "                </ASSEMBLY>\n"
            + "                <ASSEMBLY>\n"
            + "                    <STANDARD accession=\"GCA_012346.1\"/>\n"
            + "                </ASSEMBLY>\n"
            + "                <ASSEMBLY>\n"
            + "                    <STANDARD accession=\"GCA_012347.1\"/>\n"
            + "                </ASSEMBLY>\n"
            + "            </ASSEMBLY_GRAPH>\n"
            + "        </ANALYSIS_TYPE>\n"
            + "        <FILES>\n"
            + "            <FILE filename=\"assembly_graph.gfa.gz\" filetype=\"gfa\" checksum_method=\"MD5\"\n"
            + "                  checksum=\"9f2976d079c10b111669b32590d1eb3e\"/>\n"
            + "            <FILE filename=\"assembly_graph.gfa.gz\" filetype=\"gfa\" checksum_method=\"MD5\"\n"
            + "                  checksum=\"9f2976d079c10b111669b32590d1eb3e\"/>\n"
            + "        </FILES>\n"
            + "        <ANALYSIS_ATTRIBUTES>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>pangenome name</TAG>\n"
            + "                <VALUE>hprc-v1.0-mc-chm13-maxdel.10mb</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>file format version</TAG>\n"
            + "                <VALUE>GFAv1.1</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>pipeline</TAG>\n"
            + "                <VALUE>PGGB</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>resolution</TAG>\n"
            + "                <VALUE>base level</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>number of nodes</TAG>\n"
            + "                <VALUE>424643</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>number of components</TAG>\n"
            + "                <VALUE>25</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>number of edges</TAG>\n"
            + "                <VALUE>637628</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>length (bp)</TAG>\n"
            + "                <VALUE>3239764787</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>total path length (bp)</TAG>\n"
            + "                <VALUE>3239764787</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>reference assembly</TAG>\n"
            + "                <VALUE>GRCh38</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>reference assembly</TAG>\n"
            + "                <VALUE>ABCD123</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>reference assembly</TAG>\n"
            + "                <VALUE>ABCD234</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "            <ANALYSIS_ATTRIBUTE>\n"
            + "                <TAG>reference based</TAG>\n"
            + "                <VALUE>true</VALUE>\n"
            + "            </ANALYSIS_ATTRIBUTE>\n"
            + "        </ANALYSIS_ATTRIBUTES>\n"
            + "    </ANALYSIS>\n"
            + "</ANALYSIS_SET>\n";

    assertJson(
        "{\n"
            + "  \"alias\" : \"graph_X_PGGB\",\n"
            + "  \"title\" : \"Pangenome assembly graph X - generated using PGGB pipeline\",\n"
            + "  \"description\" : \"Pangenome assembly graph X was generated . Etc. \",\n"
            + "  \"study\" : {\n"
            + "    \"accession\" : \"ERP123456\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"accession\" : \"ERS123456\"\n"
            + "  }, {\n"
            + "    \"accession\" : \"ERS123457\"\n"
            + "  }, {\n"
            + "    \"accession\" : \"ERS123458\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"ASSEMBLY_GRAPH\",\n"
            + "  \"assemblies\" : [ {\n"
            + "    \"assembly\" : {\n"
            + "      \"accession\" : \"GCA_012345.1\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"assembly\" : {\n"
            + "      \"accession\" : \"GCA_012346.1\"\n"
            + "    }\n"
            + "  }, {\n"
            + "    \"assembly\" : {\n"
            + "      \"accession\" : \"GCA_012347.1\"\n"
            + "    }\n"
            + "  } ],\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"gfa\",\n"
            + "    \"checksum\" : \"9f2976d079c10b111669b32590d1eb3e\",\n"
            + "    \"fileName\" : \"assembly_graph.gfa.gz\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  }, {\n"
            + "    \"fileType\" : \"gfa\",\n"
            + "    \"checksum\" : \"9f2976d079c10b111669b32590d1eb3e\",\n"
            + "    \"fileName\" : \"assembly_graph.gfa.gz\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"pangenome name\",\n"
            + "    \"value\" : \"hprc-v1.0-mc-chm13-maxdel.10mb\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"file format version\",\n"
            + "    \"value\" : \"GFAv1.1\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"pipeline\",\n"
            + "    \"value\" : \"PGGB\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"resolution\",\n"
            + "    \"value\" : \"base level\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"number of nodes\",\n"
            + "    \"value\" : \"424643\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"number of components\",\n"
            + "    \"value\" : \"25\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"number of edges\",\n"
            + "    \"value\" : \"637628\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"length (bp)\",\n"
            + "    \"value\" : \"3239764787\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"total path length (bp)\",\n"
            + "    \"value\" : \"3239764787\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"reference assembly\",\n"
            + "    \"value\" : \"GRCh38\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"reference assembly\",\n"
            + "    \"value\" : \"ABCD123\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"reference assembly\",\n"
            + "    \"value\" : \"ABCD234\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"reference based\",\n"
            + "    \"value\" : \"true\"\n"
            + "  } ]\n"
            + "}",
        enaSraConverter.convertAnalysisXmlToJson(xml));
  }

  @Test
  /*
  SEQUENCE_VARIATION analysis type has the ASSEMBLY complex type and key value pairs, hence testing this mixed type
   */
  public void testAnalysisXml_SequenceVariationAnalysisType_ConversionToJson()
      throws JSONException {
    String xml =
        "<ANALYSIS_SET>\n"
            + "    <ANALYSIS alias=\"E-SYBR-1:F17999\" center_name=\"EMBL European Bioinformatics Institute\" broker_name=\"ArrayExpress\" accession=\"ERZ009618\">\n"
            + "        <IDENTIFIERS>\n"
            + "            <PRIMARY_ID>ERZ009618</PRIMARY_ID>\n"
            + "            <SUBMITTER_ID namespace=\"EMBL European Bioinformatics Institute\">E-SYBR-1:F17999</SUBMITTER_ID>\n"
            + "        </IDENTIFIERS>\n"
            + "        <TITLE>Multiazole resistance in clinical and environmental fungal isolates</TITLE>\n"
            + "        <DESCRIPTION>bowtie2 aligner was used to align fastq files to the reference genome and produced sam files. samtools mpileup was used for variant calling. We filtered bcf files as described in Fedorova, N., et al. (2011). Command line \"samtools mpileup -E -ugf .fa .bam | bcftools view -bvcg - > .bcf\" followed by \"bcftools view .bcf | vcfutils.pl varFilter -D 500 -d 20 | ./filter.pl > .vcf\"</DESCRIPTION>\n"
            + "        <STUDY_REF refname=\"E-SYBR-1\" refcenter=\"EMBL European Bioinformatics Institute\" accession=\"ERP002270\">\n"
            + "            <IDENTIFIERS>\n"
            + "                <PRIMARY_ID>ERP002270</PRIMARY_ID>\n"
            + "                <SUBMITTER_ID namespace=\"EMBL European Bioinformatics Institute\">E-SYBR-1</SUBMITTER_ID>\n"
            + "            </IDENTIFIERS>\n"
            + "        </STUDY_REF>\n"
            + "        <SAMPLE_REF refname=\"E-SYBR-1:F17999\" refcenter=\"EMBL European Bioinformatics Institute\" accession=\"ERS216917\">\n"
            + "            <IDENTIFIERS>\n"
            + "                <PRIMARY_ID>ERS216917</PRIMARY_ID>\n"
            + "                <SUBMITTER_ID namespace=\"EMBL European Bioinformatics Institute\">E-SYBR-1:F17999</SUBMITTER_ID>\n"
            + "            </IDENTIFIERS>\n"
            + "        </SAMPLE_REF>\n"
            + "        <ANALYSIS_TYPE>\n"
            + "            <SEQUENCE_VARIATION>\n"
            + "                <ASSEMBLY>\n"
            + "                    <STANDARD refname=\"ASM265v1\" accession=\"GCA_000002655\"/>\n"
            + "                </ASSEMBLY>\n"
            + "                <PROGRAM>test_genome_map_program</PROGRAM>\n"
            + "                <PLATFORM>BioNano</PLATFORM>\n"
            + "            </SEQUENCE_VARIATION>\n"
            + "        </ANALYSIS_TYPE>\n"
            + "        <FILES>\n"
            + "            <FILE filename=\"ERZ009/ERZ009618/C023MABXX_3_1.vcf\" filetype=\"vcf\" checksum_method=\"MD5\" checksum=\"ce5b339485e8f7e8f5d9d1dc4ef67d48\"/>\n"
            + "        </FILES>\n"
            + "    </ANALYSIS>\n"
            + "</ANALYSIS_SET>\n";

    assertJson(
        "{\n"
            + "  \"alias\" : \"E-SYBR-1:F17999\",\n"
            + "  \"accession\" : \"ERZ009618\",\n"
            + "  \"identifiers\" : { },\n"
            + "  \"centerName\" : \"EMBL European Bioinformatics Institute\",\n"
            + "  \"title\" : \"Multiazole resistance in clinical and environmental fungal isolates\",\n"
            + "  \"description\" : \"bowtie2 aligner was used to align fastq files to the reference genome and produced sam files. samtools mpileup was used for variant calling. We filtered bcf files as described in Fedorova, N., et al. (2011). Command line \\\"samtools mpileup -E -ugf .fa .bam | bcftools view -bvcg - > .bcf\\\" followed by \\\"bcftools view .bcf | vcfutils.pl varFilter -D 500 -d 20 | ./filter.pl > .vcf\\\"\",\n"
            + "  \"study\" : {\n"
            + "    \"accession\" : \"ERP002270\",\n"
            + "    \"alias\" : \"E-SYBR-1\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"accession\" : \"ERS216917\",\n"
            + "    \"alias\" : \"E-SYBR-1:F17999\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"SEQUENCE_VARIATION\",\n"
            + "  \"assemblies\" : [ {\n"
            + "    \"assembly\" : {\n"
            + "      \"accession\" : \"GCA_000002655\",\n"
            + "      \"alias\" : \"ASM265v1\"\n"
            + "    }\n"
            + "  } ],\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"vcf\",\n"
            + "    \"checksum\" : \"ce5b339485e8f7e8f5d9d1dc4ef67d48\",\n"
            + "    \"fileName\" : \"ERZ009/ERZ009618/C023MABXX_3_1.vcf\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"PLATFORM\",\n"
            + "    \"value\" : \"BioNano\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"PROGRAM\",\n"
            + "    \"value\" : \"test_genome_map_program\"\n"
            + "  } ]\n"
            + "}\n",
        enaSraConverter.convertAnalysisXmlToJson(xml));
  }

  @Test
  /*REFERENCE_ALIGNMENT analysis type used with no child elements at all*/
  public void testAnalysisXml_ReferenceAlignmentAnalysisType_ConversionToJson()
      throws JSONException {
    String xml =
        "<ANALYSIS_SET>\n"
            + "    <ANALYSIS alias=\"\">\n"
            + "        <TITLE>Test</TITLE>\n"
            + "        <DESCRIPTION>Test</DESCRIPTION>\n"
            + "        <STUDY_REF refname=\"\"/>\n"
            + "        <SAMPLE_REF refname=\"\"/>\n"
            + "        <ANALYSIS_TYPE>\n"
            + "            <REFERENCE_ALIGNMENT/>\n"
            + "        </ANALYSIS_TYPE>\n"
            + "        <FILES>\n"
            + "            <FILE filename=\"Bmi1_bw.bam\" filetype=\"bam\" checksum_method=\"MD5\"\n"
            + "                  checksum=\"9840f585055afc37de353706fd31a377\"/>\n"
            + "        </FILES>\n"
            + "    </ANALYSIS>\n"
            + "</ANALYSIS_SET>\n";

    assertJson(
        "{\n"
            + "  \"alias\" : \"\",\n"
            + "  \"title\" : \"Test\",\n"
            + "  \"description\" : \"Test\",\n"
            + "  \"study\" : {\n"
            + "    \"alias\" : \"\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"alias\" : \"\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"REFERENCE_ALIGNMENT\",\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"bam\",\n"
            + "    \"checksum\" : \"9840f585055afc37de353706fd31a377\",\n"
            + "    \"fileName\" : \"Bmi1_bw.bam\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ ]"
            + "}\n",
        enaSraConverter.convertAnalysisXmlToJson(xml));
  }

  @Test
  public void testAnalysisXml_ReferenceAlignmentSequenceAccessions_ConversionToJson()
      throws JSONException {
    String xml =
        "<ANALYSIS_SET>\n"
            + "    <ANALYSIS alias=\"\">\n"
            + "        <TITLE>Test</TITLE>\n"
            + "        <DESCRIPTION>Test</DESCRIPTION>\n"
            + "        <STUDY_REF refname=\"\"/>\n"
            + "        <SAMPLE_REF refname=\"\"/>\n"
            + "        <ANALYSIS_TYPE>\n"
            + "            <REFERENCE_ALIGNMENT>\n"
            + "                <ASSEMBLY>\n"
            + "                    <STANDARD accession=\"ADAS00000000.1\"/>\n"
            + "                </ASSEMBLY>\n"
            + "                <SEQUENCE accession=\"GG705407.1\" label=\"GG705407\"/>\n"
            + "                <SEQUENCE accession=\"GG705408.1\" label=\"GG705408\"/>\n"
            + "                <SEQUENCE accession=\"GG705409.1\" label=\"GG705409\"/>"
            + "            </REFERENCE_ALIGNMENT>\n"
            + "        </ANALYSIS_TYPE>\n"
            + "        <FILES>\n"
            + "            <FILE filename=\"Bmi1_bw.bam\" filetype=\"bam\" checksum_method=\"MD5\"\n"
            + "                  checksum=\"9840f585055afc37de353706fd31a377\"/>\n"
            + "        </FILES>\n"
            + "    </ANALYSIS>\n"
            + "</ANALYSIS_SET>";

    assertJson(
        "{\n"
            + "  \"alias\" : \"\",\n"
            + "  \"title\" : \"Test\",\n"
            + "  \"description\" : \"Test\",\n"
            + "  \"study\" : {\n"
            + "    \"alias\" : \"\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"alias\" : \"\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"REFERENCE_ALIGNMENT\",\n"
            + "  \"assemblies\": [\n"
            + "      {\n"
            + "        \"assembly\": {\n"
            + "          \"accession\": \"ADAS00000000.1\"\n"
            + "        }\n"
            + "      }\n"
            + "    ],"
            + "  \"sequences\": ["
            + "    {"
            + "      \"accession\": \"GG705407.1\","
            + "      \"label\": \"GG705407\""
            + "    },"
            + "    {"
            + "      \"accession\": \"GG705408.1\","
            + "      \"label\": \"GG705408\""
            + "    },"
            + "    {"
            + "      \"accession\": \"GG705409.1\","
            + "      \"label\": \"GG705409\""
            + "    }"
            + "  ],"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"bam\",\n"
            + "    \"checksum\" : \"9840f585055afc37de353706fd31a377\",\n"
            + "    \"fileName\" : \"Bmi1_bw.bam\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : []"
            + "}",
        enaSraConverter.convertAnalysisXmlToJson(xml));
  }

  @Test
  public void testAnalysisXml_SequenceVariationAnalysisType_WithExperimentType_ConversionToJson()
      throws JSONException {
    String xml =
        "<ANALYSIS_SET>\n"
            + "                        <ANALYSIS alias=\"E-SYBR-1:F17999\" center_name=\"EMBL European Bioinformatics Institute\" broker_name=\"ArrayExpress\" accession=\"ERZ009618\">\n"
            + "                            <IDENTIFIERS>\n"
            + "                                <PRIMARY_ID>ERZ009618</PRIMARY_ID>\n"
            + "                                <SUBMITTER_ID namespace=\"EMBL European Bioinformatics Institute\">E-SYBR-1:F17999</SUBMITTER_ID>\n"
            + "                            </IDENTIFIERS>\n"
            + "                            <TITLE>Multiazole resistance in clinical and environmental fungal isolates</TITLE>\n"
            + "                            <DESCRIPTION>bowtie2 aligner was used to align fastq files to the reference genome and produced sam files. samtools mpileup was used for variant calling. We filtered bcf files as described in Fedorova, N., et al. (2011). Command line \"samtools mpileup -E -ugf .fa .bam | bcftools view -bvcg - > .bcf\" followed by \"bcftools view .bcf | vcfutils.pl varFilter -D 500 -d 20 | ./filter.pl > .vcf\"</DESCRIPTION>\n"
            + "                            <STUDY_REF refname=\"E-SYBR-1\" refcenter=\"EMBL European Bioinformatics Institute\" accession=\"ERP002270\">\n"
            + "                                <IDENTIFIERS>\n"
            + "                                    <PRIMARY_ID>ERP002270</PRIMARY_ID>\n"
            + "                                    <SUBMITTER_ID namespace=\"EMBL European Bioinformatics Institute\">E-SYBR-1</SUBMITTER_ID>\n"
            + "                                </IDENTIFIERS>\n"
            + "                            </STUDY_REF>\n"
            + "                            <SAMPLE_REF refname=\"E-SYBR-1:F17999\" refcenter=\"EMBL European Bioinformatics Institute\" accession=\"ERS216917\">\n"
            + "                                <IDENTIFIERS>\n"
            + "                                    <PRIMARY_ID>ERS216917</PRIMARY_ID>\n"
            + "                                    <SUBMITTER_ID namespace=\"EMBL European Bioinformatics Institute\">E-SYBR-1:F17999</SUBMITTER_ID>\n"
            + "                                </IDENTIFIERS>\n"
            + "                            </SAMPLE_REF>\n"
            + "                            <ANALYSIS_TYPE>\n"
            + "                                <SEQUENCE_VARIATION>\n"
            + "                                    <ASSEMBLY>\n"
            + "                                        <STANDARD refname=\"ASM265v1\" accession=\"GCA_000002655\"/>\n"
            + "                                    </ASSEMBLY>\n"
            + "                                    <PROGRAM>test_genome_map_program</PROGRAM>\n"
            + "                                    <PLATFORM>BioNano</PLATFORM>\n"
            + "                                    <EXPERIMENT_TYPE>Whole genome sequencing</EXPERIMENT_TYPE>\n"
            + "                                    <EXPERIMENT_TYPE>Exome sequencing</EXPERIMENT_TYPE>\n"
            + "                                </SEQUENCE_VARIATION>\n"
            + "                            </ANALYSIS_TYPE>\n"
            + "                            <FILES>\n"
            + "                                <FILE filename=\"ERZ009/ERZ009618/C023MABXX_3_1.vcf\" filetype=\"vcf\" checksum_method=\"MD5\" checksum=\"ce5b339485e8f7e8f5d9d1dc4ef67d48\"/>\n"
            + "                            </FILES>\n"
            + "                        </ANALYSIS>\n"
            + "                    </ANALYSIS_SET>";

    assertJson(
        "{\n"
            + "  \"alias\" : \"E-SYBR-1:F17999\",\n"
            + "  \"accession\" : \"ERZ009618\",\n"
            + "  \"identifiers\" : { },\n"
            + "  \"centerName\" : \"EMBL European Bioinformatics Institute\",\n"
            + "  \"title\" : \"Multiazole resistance in clinical and environmental fungal isolates\",\n"
            + "  \"description\" : \"bowtie2 aligner was used to align fastq files to the reference genome and produced sam files. samtools mpileup was used for variant calling. We filtered bcf files as described in Fedorova, N., et al. (2011). Command line \\\"samtools mpileup -E -ugf .fa .bam | bcftools view -bvcg - > .bcf\\\" followed by \\\"bcftools view .bcf | vcfutils.pl varFilter -D 500 -d 20 | ./filter.pl > .vcf\\\"\",\n"
            + "  \"study\" : {\n"
            + "    \"accession\" : \"ERP002270\",\n"
            + "    \"alias\" : \"E-SYBR-1\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"accession\" : \"ERS216917\",\n"
            + "    \"alias\" : \"E-SYBR-1:F17999\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"SEQUENCE_VARIATION\",\n"
            + "  \"assemblies\" : [ {\n"
            + "    \"assembly\" : {\n"
            + "      \"accession\" : \"GCA_000002655\",\n"
            + "      \"alias\" : \"ASM265v1\"\n"
            + "    }\n"
            + "  } ],\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"vcf\",\n"
            + "    \"checksum\" : \"ce5b339485e8f7e8f5d9d1dc4ef67d48\",\n"
            + "    \"fileName\" : \"ERZ009/ERZ009618/C023MABXX_3_1.vcf\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"EXPERIMENT_TYPE\",\n"
            + "    \"value\" : \"Whole genome sequencing\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"EXPERIMENT_TYPE\",\n"
            + "    \"value\" : \"Exome sequencing\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"PLATFORM\",\n"
            + "    \"value\" : \"BioNano\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"PROGRAM\",\n"
            + "    \"value\" : \"test_genome_map_program\"\n"
            + "  } ]\n"
            + "}",
        enaSraConverter.convertAnalysisXmlToJson(xml));
  }

  @Test
  public void testAnalysisJson_GenomeMapAnalysisType_ConversionToXml() {
    String json =
        "{\n"
            + "  \"alias\" : \"\",\n"
            + "  \"centerName\" : \"\",\n"
            + "  \"title\" : \"Pathogen_analysis title\",\n"
            + "  \"study\" : {\n"
            + "    \"alias\" : \"\"\n"
            + "  },\n"
            + "  \"samples\" : [ ],\n"
            + "  \"analysisType\" : \"GENOME_MAP\",\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"tab\",\n"
            + "    \"checksum\" : \"5a8018b5d369af96ccc07dc77b923fb1\",\n"
            + "    \"fileName\" : \"sequence_flatfile.tsv.gz\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  }, {\n"
            + "    \"fileType\" : \"BioNano_native\",\n"
            + "    \"checksum\" : \"16cf6f36e00ee61ba888ae14efd83873\",\n"
            + "    \"fileName\" : \"test.bnx\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  }, {\n"
            + "    \"fileType\" : \"BioNano_native\",\n"
            + "    \"checksum\" : \"16cf6f36e00ee61ba888ae14efd83873\",\n"
            + "    \"fileName\" : \"dir/test.bnx\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"PROGRAM\",\n"
            + "    \"value\" : \"test_genome_map_program\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"PLATFORM\",\n"
            + "    \"value\" : \"BioNano\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"DESCRIPTION\",\n"
            + "    \"value\" : \"test_genome_map_description\"\n"
            + "  } ]\n"
            + "}\n";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<ANALYSIS alias=\"\" center_name=\"\">\n"
            + "  <TITLE>Pathogen_analysis title</TITLE>\n"
            + "  <STUDY_REF refname=\"\"/>\n"
            + "  <ANALYSIS_TYPE>\n"
            + "    <GENOME_MAP>\n"
            + "      <PLATFORM>BioNano</PLATFORM>\n"
            + "      <DESCRIPTION>test_genome_map_description</DESCRIPTION>\n"
            + "      <PROGRAM>test_genome_map_program</PROGRAM>\n"
            + "    </GENOME_MAP>\n"
            + "  </ANALYSIS_TYPE>\n"
            + "  <FILES>\n"
            + "    <FILE filename=\"sequence_flatfile.tsv.gz\" filetype=\"tab\" checksum_method=\"MD5\" checksum=\"5a8018b5d369af96ccc07dc77b923fb1\"/>\n"
            + "    <FILE filename=\"test.bnx\" filetype=\"BioNano_native\" checksum_method=\"MD5\" checksum=\"16cf6f36e00ee61ba888ae14efd83873\"/>\n"
            + "    <FILE filename=\"dir/test.bnx\" filetype=\"BioNano_native\" checksum_method=\"MD5\" checksum=\"16cf6f36e00ee61ba888ae14efd83873\"/>\n"
            + "  </FILES>\n"
            + "  <ANALYSIS_ATTRIBUTES/>\n"
            + "</ANALYSIS>",
        enaSraConverter.convertAnalysisJsonToXml(json));
  }

  @Test
  public void testAnalysisJson_SequenceVariationAnalysisType_ConversionToXml() {
    String json =
        "{\n"
            + "  \"alias\" : \"E-SYBR-1:F17999\",\n"
            + "  \"accession\" : \"ERZ009618\",\n"
            + "  \"identifiers\" : { },\n"
            + "  \"centerName\" : \"EMBL European Bioinformatics Institute\",\n"
            + "  \"title\" : \"Multiazole resistance in clinical and environmental fungal isolates\",\n"
            + "  \"description\" : \"bowtie2 aligner was used to align fastq files to the reference genome and produced sam files. samtools mpileup was used for variant calling. We filtered bcf files as described in Fedorova, N., et al. (2011). Command line \\\"samtools mpileup -E -ugf .fa .bam | bcftools view -bvcg - > .bcf\\\" followed by \\\"bcftools view .bcf | vcfutils.pl varFilter -D 500 -d 20 | ./filter.pl > .vcf\\\"\",\n"
            + "  \"study\" : {\n"
            + "    \"accession\" : \"ERP002270\",\n"
            + "    \"alias\" : \"E-SYBR-1\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"accession\" : \"ERS216917\",\n"
            + "    \"alias\" : \"E-SYBR-1:F17999\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"SEQUENCE_VARIATION\",\n"
            + "  \"assemblies\" : [ {\n"
            + "    \"assembly\" : {\n"
            + "      \"accession\" : \"GCA_000002655\",\n"
            + "      \"alias\" : \"ASM265v1\"\n"
            + "    }\n"
            + "  } ],\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"vcf\",\n"
            + "    \"checksum\" : \"ce5b339485e8f7e8f5d9d1dc4ef67d48\",\n"
            + "    \"fileName\" : \"ERZ009/ERZ009618/C023MABXX_3_1.vcf\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ],\n"
            + "  \"attributes\" : [ {\n"
            + "    \"tag\" : \"PLATFORM\",\n"
            + "    \"value\" : \"BioNano\"\n"
            + "  }, {\n"
            + "    \"tag\" : \"PROGRAM\",\n"
            + "    \"value\" : \"test_genome_map_program\"\n"
            + "  } ]\n"
            + "}\n";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<ANALYSIS alias=\"E-SYBR-1:F17999\" accession=\"ERZ009618\" center_name=\"EMBL European Bioinformatics Institute\">\n"
            + "  <IDENTIFIERS/>\n"
            + "  <TITLE>Multiazole resistance in clinical and environmental fungal isolates</TITLE>\n"
            + "  <DESCRIPTION>bowtie2 aligner was used to align fastq files to the reference genome and produced sam files. samtools mpileup was used for variant calling. We filtered bcf files as described in Fedorova, N., et al. (2011). Command line \"samtools mpileup -E -ugf .fa .bam | bcftools view -bvcg - > .bcf\" followed by \"bcftools view .bcf | vcfutils.pl varFilter -D 500 -d 20 | ./filter.pl > .vcf\"</DESCRIPTION>\n"
            + "  <STUDY_REF accession=\"ERP002270\" refname=\"E-SYBR-1\"/>\n"
            + "  <SAMPLE_REF accession=\"ERS216917\" refname=\"E-SYBR-1:F17999\"/>\n"
            + "  <ANALYSIS_TYPE>\n"
            + "    <SEQUENCE_VARIATION>\n"
            + "      <ASSEMBLY>\n"
            + "        <STANDARD accession=\"GCA_000002655\" refname=\"ASM265v1\"/>\n"
            + "      </ASSEMBLY>\n"
            + "      <PLATFORM>BioNano</PLATFORM>\n"
            + "      <PROGRAM>test_genome_map_program</PROGRAM>\n"
            + "    </SEQUENCE_VARIATION>\n"
            + "  </ANALYSIS_TYPE>\n"
            + "  <FILES>\n"
            + "    <FILE filename=\"ERZ009/ERZ009618/C023MABXX_3_1.vcf\" filetype=\"vcf\" checksum_method=\"MD5\" checksum=\"ce5b339485e8f7e8f5d9d1dc4ef67d48\"/>\n"
            + "  </FILES>\n"
            + "  <ANALYSIS_ATTRIBUTES/>\n"
            + "</ANALYSIS>",
        enaSraConverter.convertAnalysisJsonToXml(json));
  }

  @Test
  public void testAnalysisJson_ReferenceAlignmentAnalysisType_ConversionToXml() {
    String json =
        "{\n"
            + "  \"alias\" : \"\",\n"
            + "  \"title\" : \"Test\",\n"
            + "  \"description\" : \"Test\",\n"
            + "  \"study\" : {\n"
            + "    \"alias\" : \"\"\n"
            + "  },\n"
            + "  \"samples\" : [ {\n"
            + "    \"alias\" : \"\"\n"
            + "  } ],\n"
            + "  \"analysisType\" : \"REFERENCE_ALIGNMENT\",\n"
            + "  \"files\" : [ {\n"
            + "    \"fileType\" : \"bam\",\n"
            + "    \"checksum\" : \"9840f585055afc37de353706fd31a377\",\n"
            + "    \"fileName\" : \"Bmi1_bw.bam\",\n"
            + "    \"checksumMethod\" : \"MD5\"\n"
            + "  } ]\n"
            + "}\n";

    assertXml(
        "<?xml version='1.0' encoding='UTF-8'?>\n"
            + "<ANALYSIS alias=\"\">\n"
            + "  <TITLE>Test</TITLE>\n"
            + "  <DESCRIPTION>Test</DESCRIPTION>\n"
            + "  <STUDY_REF refname=\"\"/>\n"
            + "  <SAMPLE_REF refname=\"\"/>\n"
            + "  <ANALYSIS_TYPE>\n"
            + "    <REFERENCE_ALIGNMENT/>\n"
            + "  </ANALYSIS_TYPE>\n"
            + "  <FILES>\n"
            + "    <FILE filename=\"Bmi1_bw.bam\" filetype=\"bam\" checksum_method=\"MD5\" checksum=\"9840f585055afc37de353706fd31a377\"/>\n"
            + "  </FILES>\n"
            + "  <ANALYSIS_ATTRIBUTES/>\n"
            + "</ANALYSIS>\n",
        enaSraConverter.convertAnalysisJsonToXml(json));
  }

  /**
   * Asserts that the expected Receipt XML matches the actual Receipt XML. Ignores accession and
   * receipt date.
   */
  private void assertReceiptXml(String expected, String actual) {
    actual = actual.replaceAll("\\<\\?xml.*\\n", "");
    Diff diff =
        DiffBuilder.compare(expected)
            .withDifferenceEvaluator(
                DifferenceEvaluators.chain(
                    DifferenceEvaluators.Default,
                    DifferenceEvaluators.downgradeDifferencesToEqual(
                        ComparisonType.XML_VERSION,
                        ComparisonType.XML_ENCODING,
                        ComparisonType.XML_STANDALONE)))
            .withTest(actual)
            .withNodeMatcher(new DefaultNodeMatcher(ElementSelectors.byName))
            .ignoreWhitespace()
            .withAttributeFilter(
                a -> {
                  if (a.getName().equals("accession")
                      || a.getName().equals("receiptDate")
                      || a.getName().equals("submissionFile")
                      || a.getName().equals("holdUntilDate")) {
                    return false;
                  }
                  return true;
                })
            .build();
    if (diff.hasDifferences()) {
      // Show the difference.
      assertEquals(expected, actual);
    }
  }

  private void assertXml(String expected, String actual) {
    actual = actual.replaceAll("\\<\\?xml.*\\n", "");
    Diff diff =
        DiffBuilder.compare(expected)
            .withDifferenceEvaluator(
                DifferenceEvaluators.chain(
                    DifferenceEvaluators.Default,
                    DifferenceEvaluators.downgradeDifferencesToEqual(
                        ComparisonType.XML_VERSION,
                        ComparisonType.XML_ENCODING,
                        ComparisonType.XML_STANDALONE)))
            .withTest(actual)
            .withNodeMatcher(new DefaultNodeMatcher(ElementSelectors.byName))
            .ignoreWhitespace()
            .build();
    if (diff.hasDifferences()) {
      // Show the difference.
      assertEquals(expected, actual);
    }
  }

  /**
   * Asserts that the expected Receipt XML matches the actual Receipt XML. Ignores accession and
   * receipt date.
   */
  public static void assertReceiptJson(String expected, String actual) throws JSONException {
    JSONAssert.assertEquals(
        expected,
        actual,
        new CustomComparator(
            JSONCompareMode.STRICT,
            new Customization("receiptDate", (o1, o2) -> true),
            new Customization("samples[*].accession", (o1, o2) -> true),
            new Customization("samples[*].externalAccession.id", (o1, o2) -> true),
            new Customization("studies[*].accession", (o1, o2) -> true),
            new Customization("studies[*].holdUntilDate", (o1, o2) -> true),
            new Customization("projects[*].accession", (o1, o2) -> true),
            new Customization("experiments[*].accession", (o1, o2) -> true),
            new Customization("analyses[*].accession", (o1, o2) -> true),
            new Customization("runs[*].accession", (o1, o2) -> true),
            new Customization("projects[*].holdUntilDate", (o1, o2) -> true),
            new Customization("studies[*].externalAccession.id", (o1, o2) -> true),
            new Customization("submission.accession", (o1, o2) -> true),
            new Customization("id", (o1, o2) -> true)));
  }

  public static void assertJson(String expected, String actual) throws JSONException {
    try {
      JSONAssert.assertEquals(expected, actual, JSONCompareMode.STRICT);
    } catch (Throwable ex) {
      ex.printStackTrace();

      throw new ComparisonFailure("JSON comparison failed.", expected, actual);
    }
  }
}
